{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The function to get the reverse complementary sequences\n",
    "rc = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}\n",
    "def revcomp(tmp_seq):\n",
    "    return ''.join([rc[x] for x in tmp_seq[::-1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The function to read sequences from a FASTA file\n",
    "def read_fasta(tmp_filename):\n",
    "    rv = dict()\n",
    "    f = open(tmp_filename, 'r')\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            tmp_h = line.strip().lstrip('>')\n",
    "            rv[tmp_h] = ''\n",
    "        else:\n",
    "            rv[tmp_h] += line.strip().replace(' ', '')\n",
    "    f.close()\n",
    "    return rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2019-nCoV_N1-F GACCCCAAAATCAGCGAAAT\n",
      "2019-nCoV_N1-R TCTGGTTACTGCCAGTTGAATCTG\n",
      "2019-nCoV_N1-P ACCCCGCATTACGTTTGGTGGACC\n",
      "2019-nCoV_N2-F TTACAAACATTGGCCGCAAA\n",
      "2019-nCoV_N2-R GCGCGACATTCCGAAGAA\n",
      "2019-nCoV_N2-P ACAATTTGCCCCCAGCGCTTCAG\n",
      "2019-nCoV_N3-F GGGAGCCTTGAATACACCAAAA\n",
      "2019-nCoV_N3-R TGTAGCACGATTGCAGCATTG\n",
      "2019-nCoV_N3-P AYCACATTGGCACCCGCAATCCTG\n",
      "RP-F AGATTTGGACCTGCGAGCG\n",
      "RP-R GAGCGGCTGTCTCCACAAGT\n",
      "RP-P TTCTGACCTGAAGGCTCTGCGCG\n"
     ]
    }
   ],
   "source": [
    "# Read primers\n",
    "N1,N2,N3,RP={},{},{},{}\n",
    "prim=(N1,N2,N3,RP)\n",
    "filename_primers = '../data/2019nCoV_primers.fa'\n",
    "primer_list = read_fasta(filename_primers)\n",
    "for p_h, p_seq in primer_list.items():\n",
    "    print(p_h, p_seq)\n",
    "    if p_h.find('N1')!=-1:\n",
    "        N1[p_h[-1]]=p_seq\n",
    "    elif p_h.find('N2')!=-1:\n",
    "        N2[p_h[-1]]=p_seq\n",
    "    elif p_h.find('N3')!=-1:\n",
    "        N3[p_h[-1]]=p_seq\n",
    "    else:\n",
    "        RP[p_h[-1]]=p_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENA|MN988668|MN988668.1 TTAAAGGTTTATACCTTCCCAGGTA\n",
      "ENA|MN988669|MN988669.1 TTAAAGGTTTATACCTTCCCAGGTA\n",
      "ENA|MN938384|MN938384.1 CAACCAACTTTCGATCTCTTGTAGA\n",
      "ENA|MN975262|MN975262.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN996527|MN996527.1 AACCAACTTTCGATCTCTTGTAGAT\n",
      "ENA|MN996528|MN996528.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN996529|MN996529.1 TACCTTCCCAGGTAACAAACCAACC\n",
      "ENA|MN996530|MN996530.1 CCTTCCCAGGTAACAAACCAACCAA\n",
      "ENA|MN996531|MN996531.1 ACCTTCCCAGGTAACAAACCAACCA\n",
      "ENA|MN908947|MN908947.3 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN985325|MN985325.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN988713|MN988713.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN994467|MN994467.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN994468|MN994468.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN997409|MN997409.1 ATTAAAGGTTTATACCTTCCCAGGT\n"
     ]
    }
   ],
   "source": [
    "# Read genomes\n",
    "filename_genomes = '../data/2019nCoV_genomes.2020_02_03.fa'\n",
    "genome_list = read_fasta(filename_genomes)\n",
    "for g_h, g_seq in genome_list.items():\n",
    "    print(g_h.split()[0], g_seq[:25])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N1 - 72.0 GACCCCAAAATCAGCGAAATGCACCCCGCATTACGTTTGGTGGACCCTCAGATTCAACTGGCAGTAACCAGA\n",
      "They are all same among the genomes\n",
      "N2 - 67.0 TTACAAACATTGGCCGCAAATTGCACAATTTGCCCCCAGCGCTTCAGCGTTCTTCGGAATGTCGCGC\n",
      "They are all same among the genomes\n",
      "N3 - 72.0 GGGAGCCTTGAATACACCAAAAGATCACATTGGCACCCGCAATCCTGCTAACAATGCTGCAATCGTGCTACA\n",
      "They are all same among the genomes\n",
      "RP - No amplicons\n"
     ]
    }
   ],
   "source": [
    "# Q1. Find primers in the genome\n",
    "# What is the length of amplicons generated by primers? Are they all same among the genomes?\n",
    "primers=['N1','N2','N3','RP']\n",
    "cnt=0\n",
    "TF=True\n",
    "for i in prim: #N1,N2,N3,RP\n",
    "    amp_list=[]\n",
    "    num=0\n",
    "    for j in genome_list: #genome 1~15\n",
    "        if(i['F'] in genome_list[j] and revcomp(i['R']) in genome_list[j]):\n",
    "            F,revR=genome_list[j].find(i['F']),genome_list[j].find(revcomp(i['R']))\n",
    "            amp=genome_list[j][F:revR+len(i['R'])]\n",
    "            amp_list.extend(amp)\n",
    "            if num!=0:\n",
    "                if temp_amp!=amp:\n",
    "                    TF=False\n",
    "            temp_amp=amp\n",
    "            num+=1\n",
    "    if num!=0:\n",
    "        print(primers[cnt], \"-\", len(amp_list)/num, amp)\n",
    "        if TF==True:\n",
    "            print(\"They are all same among the genomes\")\n",
    "        else:\n",
    "            print(\"They are NOT all same among the genomes\")\n",
    "    else:\n",
    "        print(primers[cnt], \"- No amplicons\")\n",
    "    cnt+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N1 primers can detect all SARS-CoV-2 genomes\n",
      "N2 primers can detect all SARS-CoV-2 genomes\n",
      "N3 primers can detect all SARS-CoV-2 genomes\n",
      "RP primers can NOT detect all SARS-CoV-2 gnomes\n"
     ]
    }
   ],
   "source": [
    "# Q2. Compare the number of amplicons and the number of genomes\n",
    "# Can these primers detect all SARS-CoV-2 genomes?\n",
    "primers=['N1','N2','N3','RP']\n",
    "cnt,same=0,0\n",
    "for i in prim: #N1,N2,N3,RP\n",
    "    for j in genome_list: #genome 1~15\n",
    "        if(i['F'] in genome_list[j] and revcomp(i['R']) in genome_list[j]):\n",
    "            same+=1\n",
    "    if same==len(genome_list):\n",
    "        print(primers[cnt], \"primers can detect all SARS-CoV-2 genomes\")\n",
    "    else:\n",
    "        print(primers[cnt], \"primers can NOT detect all SARS-CoV-2 gnomes\")\n",
    "    same=0\n",
    "    cnt+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENA|MN988668|MN988668.1 TTAAAGGTTTATACCTTCCCAGGTA\n",
      "ENA|MN988669|MN988669.1 TTAAAGGTTTATACCTTCCCAGGTA\n",
      "ENA|MN938384|MN938384.1 CAACCAACTTTCGATCTCTTGTAGA\n",
      "ENA|MN975262|MN975262.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN996527|MN996527.1 AACCAACTTTCGATCTCTTGTAGAT\n",
      "ENA|MN996528|MN996528.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN996529|MN996529.1 TACCTTCCCAGGTAACAAACCAACC\n",
      "ENA|MN996530|MN996530.1 CCTTCCCAGGTAACAAACCAACCAA\n",
      "ENA|MN996531|MN996531.1 ACCTTCCCAGGTAACAAACCAACCA\n",
      "ENA|MN908947|MN908947.3 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN985325|MN985325.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN988713|MN988713.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN994467|MN994467.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN994468|MN994468.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN997409|MN997409.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "\n",
      "\n",
      "ENA|MN988668|MN988668.1 TTAAAGGTTTATACCTTCCCAGGTA\n",
      "ENA|MN988669|MN988669.1 TTAAAGGTTTATACCTTCCCAGGTA\n",
      "ENA|MN938384|MN938384.1 CAACCAACTTTCGATCTCTTGTAGA\n",
      "ENA|MN975262|MN975262.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN996527|MN996527.1 AACCAACTTTCGATCTCTTGTAGAT\n",
      "ENA|MN996528|MN996528.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN996529|MN996529.1 TACCTTCCCAGGTAACAAACCAACC\n",
      "ENA|MN996530|MN996530.1 CCTTCCCAGGTAACAAACCAACCAA\n",
      "ENA|MN996531|MN996531.1 ACCTTCCCAGGTAACAAACCAACCA\n",
      "ENA|MN908947|MN908947.3 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN985325|MN985325.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN988713|MN988713.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN994467|MN994467.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN994468|MN994468.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN997409|MN997409.1 ATTAAAGGTTTATACCTTCCCAGGT\n"
     ]
    }
   ],
   "source": [
    "# Q3. First, Read MERS and SARS genome fasta file like 2019nCoV\n",
    "# Read MERS, SARS\n",
    "filename_MERS = '../data/MERS_genomes.2020_02_03.fa'\n",
    "filename_SARS = '../data/SARS_genomes.2020_02_03.fa'\n",
    "MERS_list = read_fasta(filename_MERS)\n",
    "for m_h, m_seq in genome_list.items():\n",
    "    print(m_h.split()[0], m_seq[:25])\n",
    "print('\\n')\n",
    "SARS_list = read_fasta(filename_SARS)\n",
    "for s_h, s_seq in genome_list.items():\n",
    "    print(s_h.split()[0], s_seq[:25])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N1 can NOT detect any MERS or SARS genomes\n",
      "N2 can NOT detect any MERS or SARS genomes\n",
      "N3 can NOT detect any MERS or SARS genomes\n",
      "RP can NOT detect any MERS or SARS genomes\n"
     ]
    }
   ],
   "source": [
    "# Q3. Proceed as in Q2\n",
    "# 3. Can these primers detect MERS genomes? How about SARS genomes?\n",
    "primers=['N1','N2','N3','RP']\n",
    "cnt=0\n",
    "MTF,STF=False,False\n",
    "for i in prim: #N1,N2,N3,RP\n",
    "    for j in MERS_list:\n",
    "        if(i['F'] in MERS_list[j] and revcomp(i['R']) in MERS_list[j]):\n",
    "            MTF=True\n",
    "    for k in SARS_list:\n",
    "        if(i['F'] in SARS_list[k] and revcomp(i['R']) in SARS_list[k]):\n",
    "            STF=True\n",
    "    if(MTF and STF):\n",
    "        print(primers[cnt], \"can detect MERS and SARS genomes\")\n",
    "    elif MTF:\n",
    "        print(primers[cnt], \"can ONLY detect MERS genomes\")\n",
    "    elif STF:\n",
    "        print(primers[cnt], \"can ONLY detect SARS genomes\")\n",
    "    else:\n",
    "        print(primers[cnt], \"can NOT detect any MERS or SARS genomes\")\n",
    "    cnt+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For the Q4 and Q5, Extract those amplicon sequences\n",
    "amp_list={}\n",
    "primers=['N1','N2','N3','RP']\n",
    "cnt=0\n",
    "for i in prim: #N1,N2,N3,RP\n",
    "    for j in genome_list: #genome 1~15\n",
    "        if(i['F'] in genome_list[j] and revcomp(i['R']) in genome_list[j]):\n",
    "            F,revR=genome_list[j].find(i['F']),genome_list[j].find(revcomp(i['R']))\n",
    "            amp=genome_list[j][F:revR+len(i['R'])]\n",
    "            amp_list[\"Amplicon generated by \"+primers[cnt]+\" in genome \"+j]=amp\n",
    "    cnt+=1\n",
    "    \n",
    "ofile = open(\"../data/amplicon_sequences.fa\",\"w\")\n",
    "for a_h, a_seq in amp_list.items():\n",
    "    ofile.write(\">\"+a_h+\"\\n\"+a_seq+\"\\n\")\n",
    "ofile.close()\n",
    "\n",
    "# makeblastdb -dbtype nucl -in 2019nCoV_genomes.2020_03_07.fa -out 2019nCoV_genomes.2020_03_07.blastdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "They all perfectly matched well :)\n"
     ]
    }
   ],
   "source": [
    "# Q4. blastn -db 2019nCoV_genomes.2020_03_07.blastdb -query amplicon_sequences.fa -outfmt 6 -out Q4.out\n",
    "# 4. Run the BLAST to all SARS-CoV-2 genomes. Are they all perfectly matched? Is there any mismatch or gap?\n",
    "f=open('../data/Q4.out', 'r')\n",
    "TF,TFTF=True,False\n",
    "for i in f:\n",
    "    if i.split()[2]!='100.000':\n",
    "        TF=False\n",
    "        print(\"It has \"+i.split()[4]+\"mismatch and \"+i.split()[5]+\"gap in\"+i.split()[1])\n",
    "    if TF:\n",
    "        TFTF=True\n",
    "if TF:\n",
    "    print(\"They all perfectly matched well :)\")\n",
    "elif TFTF:\n",
    "    print(\"The other rest perfectly matched well :P\")\n",
    "else:\n",
    "    print(\"Nothing matched :(\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Only 75.712% of the total proteome matched perfectly :P\n"
     ]
    }
   ],
   "source": [
    "# Q5. blastx -db 2019_nCoV_proteins.2020_03_27.blastdb -query amplicon_sequences.fa -outfmt 6 -out Q5.out\n",
    "# 5. Run the BLAST to all SARS-CoV-2 proteomes. Are they all perfectly matched?\n",
    "f=open('../data/Q5.out', 'r')\n",
    "Pass,Fail=0,0\n",
    "TF,TFTF=True,False\n",
    "for i in f:\n",
    "    if i.split()[2]!='100.000':\n",
    "        TF=False\n",
    "        Fail+=1\n",
    "    else:\n",
    "        Pass+=1\n",
    "    if TF:\n",
    "        TFTF=True\n",
    "\n",
    "if TF:\n",
    "    print(\"They all perfectly matched well :)\")\n",
    "elif TFTF:\n",
    "    print(\"Only \"+str('%.3f'%(Pass/(Pass+Fail)*100))+\"% of the total proteome matched perfectly :P\")\n",
    "else:\n",
    "    print(\"Nothing matched :(\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
