{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "rc = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}\n",
    "def revcomp(tmp_seq):\n",
    "    return ''.join([rc[x] for x in tmp_seq[::-1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename_primers= '../data/2019nCoV_primers.fa'\n",
    "filename_genomes_1= '../data/2019nCoV_genomes.2020_03_27.fa'\n",
    "filename_genomes_2= '../data/MERS_genomes.2020_02_03.fa'\n",
    "filename_genomes_3= '../data/SARS_genomes.2020_02_03.fa'\n",
    "def read_fasta(tmp_filename):\n",
    "    rv = dict()\n",
    "    f = open(tmp_filename, 'r')\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            tmp_h = line.strip().lstrip('>')\n",
    "            rv[tmp_h] = ''\n",
    "        else:\n",
    "            rv[tmp_h] += line.strip().replace(' ', '')\n",
    "    f.close()\n",
    "    return rv\n",
    "genome_list_1=read_fasta(filename_genomes_1)\n",
    "genome_list_2=read_fasta(filename_genomes_2)\n",
    "genome_list_3=read_fasta(filename_genomes_3)\n",
    "genome_seq_1=list(genome_list_1.values())\n",
    "genome_seq_2=list(genome_list_2.values())\n",
    "genome_seq_3=list(genome_list_3.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "primer_name=[]\n",
    "primer_sequence=[]\n",
    "primer_list = read_fasta(filename_primers)\n",
    "for p_h, p_seq in primer_list.items():\n",
    "    primer_name.append(p_h)\n",
    "    primer_sequence.append(p_seq)\n",
    "primer_name_s=[]\n",
    "amplicons_f2019primers=[]\n",
    "for i in range(len(primer_name)):\n",
    "    primer_name_s.append(primer_name[i][:-2])\n",
    "for j in range(len(list(set(primer_name_s)))):\n",
    "    amplicons_f2019primers.append(list(set(primer_name_s))[j])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The sequence and the  length of the amplicon created by  2019-nCoV_N2 is  ['TTACAAACATTGGCCGCAAATTGCACAATTTGCCCCCAGCGCTTCAGCGTTCTTCGGAATGTCGCGC']  and  67\n",
      "2019-nCoV_N2 gives all same amplicon sequences among the genomes.\n",
      "The sequence and the  length of the amplicon created by  RP is  []  and  0\n",
      "RP doesn't recognize the genomes at all.\n",
      "The sequence and the  length of the amplicon created by  2019-nCoV_N1 is  ['GACCCCAAAATCAGCGAAATGCACCCCGCATTACGTTTGGTGGACCCTCAGATTCAACTGGCAGTAACCAGA']  and  72\n",
      "2019-nCoV_N1 gives all same amplicon sequences among the genomes.\n",
      "The sequence and the  length of the amplicon created by  2019-nCoV_N3 is  ['GGGAGCCTTGAATACACCAAAAGATCACATTGGCACCCGCAATCCTGCTAACAATGCTGCAATCGTGCTACA']  and  72\n",
      "2019-nCoV_N3 gives all same amplicon sequences among the genomes.\n"
     ]
    }
   ],
   "source": [
    "#Question 1:What is the length of amplicons generated by primers? Are they all same among the genomes?\n",
    "for p_name in amplicons_f2019primers:\n",
    "    SARS_CoV_2_amplicons.update({p_name: []})\n",
    "for genome_1 in genome_seq_1:\n",
    "    for i in range(len(primer_sequence)):\n",
    "        if primer_name[i][-1]=='F':\n",
    "            f_primer=primer_sequence[i]\n",
    "            r_primer=primer_sequence[i+1]\n",
    "            if f_primer in genome_1 and revcomp(r_primer) in genome_1:\n",
    "                F_location=genome_1.find(f_primer)\n",
    "                revR_location=genome_1.find(revcomp(r_primer))\n",
    "                amp=genome_1[F_location:revR_location+len(r_primer)]\n",
    "                SARS_CoV_2_amplicons[primer_name_s[i]].append(amp) \n",
    "for p_name in amplicons_f2019primers:\n",
    "        ab=list(set((SARS_CoV_2_amplicons[p_name])))\n",
    "        print(\"The sequence and the  length of the amplicon created by \",p_name, \"is \",ab,\" and \", len(''.join(ab)))\n",
    "        if len(ab)==1:\n",
    "            print(p_name, \"gives all same amplicon sequences among the genomes.\")\n",
    "        elif len(ab)==0:\n",
    "            print(p_name, \"doesn't recognize the genomes at all.\")\n",
    "        else: \n",
    "            print(p_name, \"There are \",len(ab),\"different numbers of amplicon result.\")            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2019-nCoV_N2  could detect all  73  cases.\n",
      "RP  couldn't detect  73  out of  73  cases.\n",
      "2019-nCoV_N1  could detect all  73  cases.\n",
      "2019-nCoV_N3  could detect all  73  cases.\n"
     ]
    }
   ],
   "source": [
    "#2. Can these primers detect all SARS-CoV-2 genomes?\n",
    "for p_name in amplicons_f2019primers:\n",
    "        ab=list(set((SARS_CoV_2_amplicons[p_name])))\n",
    "        if len(ab)==1:\n",
    "            print(p_name, \" could detect all \",len(genome_seq_1),\" cases.\")\n",
    "        elif len(ab)==0:\n",
    "            print(p_name, \" couldn't detect \", len(genome_seq_1), \" out of \",len(genome_seq_1),\" cases.\")\n",
    "        else:\n",
    "            if '' in ab:\n",
    "                print(p_name, \" couldn't detect \", ab.count(''), \" out of \",len(genome_seq_1),\" cases.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2019-nCoV_N2  can't detect MERS genomes.\n",
      "RP  can't detect MERS genomes.\n",
      "2019-nCoV_N1  can't detect MERS genomes.\n",
      "2019-nCoV_N3  can't detect MERS genomes.\n"
     ]
    }
   ],
   "source": [
    "#3. Can these primers detect MERS genomes? \n",
    "for p_name in amplicons_f2019primers:\n",
    "    MERS_amplicons.update({p_name: []})\n",
    "for genome_2 in genome_seq_2:\n",
    "    for i in range(len(primer_sequence)):\n",
    "        if primer_name[i][-1]=='F':\n",
    "            f_primer=primer_sequence[i]\n",
    "            r_primer=primer_sequence[i+1]\n",
    "            if f_primer in genome_2 and revcomp(r_primer) in genome_2:\n",
    "                F_location=genome_2.find(f_primer)\n",
    "                revR_location=genome_2.find(revcomp(r_primer))\n",
    "                amp=genome_2[F_location:revR_location+len(r_primer)]\n",
    "                MERS_amplicons[primer_name_s[i]].append(amp) \n",
    "\n",
    "for primer_name in amplicons_f2019primers:\n",
    "    ab=list(set((MERS_amplicons[p_name])))\n",
    "    if len(MERS_amplicons[primer_name])==0:\n",
    "        print(primer_name, \" can't detect MERS genomes.\")\n",
    "        continue\n",
    "    else:\n",
    "        print(primer_name, \" can detect MERS genomes.\")\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2019-nCoV_N2  can't detect SARS genomes.\n",
      "RP  can't detect SARS genomes.\n",
      "2019-nCoV_N1  can't detect SARS genomes.\n",
      "2019-nCoV_N3  can't detect SARS genomes.\n"
     ]
    }
   ],
   "source": [
    "#Q-3:How about SARS genomes?\n",
    "for p_name in amplicons_f2019primers:\n",
    "    SARS_amplicons.update({p_name: []})\n",
    "for genome_3 in genome_seq_3:\n",
    "    for i in range(len(primer_sequence)):\n",
    "        if primer_name[i][-1]=='F':\n",
    "            f_primer=primer_sequence[i]\n",
    "            r_primer=primer_sequence[i+1]\n",
    "            if f_primer in genome_3 and revcomp(r_primer) in genome_3:\n",
    "                F_location=genome_3.find(f_primer)\n",
    "                revR_location=genome_3.find(revcomp(r_primer))\n",
    "                amp=genome_3[F_location:revR_location+len(r_primer)]\n",
    "                MERS_amplicons[primer_name_s[i]].append(amp) \n",
    "\n",
    "for primer_name in amplicons_f2019primers:\n",
    "    ab=list(set((SARS_amplicons[p_name])))\n",
    "    if len(SARS_amplicons[primer_name])==0:\n",
    "        print(primer_name, \" can't detect SARS genomes.\")\n",
    "        continue\n",
    "    else:\n",
    "        print(primer_name, \" can detect SARS genomes.\")\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Q-4:extract the amplicon sequences\n",
    "ofile = open(\"20171224ampliconfastafile.fa\", \"w\")\n",
    "for p_name in amplicons_f2019primers:\n",
    "    if len(SARS_CoV_2_amplicons[p_name])!=0:\n",
    "        ab=list(set((SARS_CoV_2_amplicons[p_name])))\n",
    "        ofile.write(\">\" + p_name + \"\\n\" +''.join(ab)+\"\\n\")\n",
    "ofile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Every genomes and primers perfectly match.\n"
     ]
    }
   ],
   "source": [
    "#Q-4.Extract those amplicon sequences, and run the BLAST to all SARS-CoV-2 genomes. Are they all perfectly matched? Is there any mismatch or gap?\n",
    "Genome_blast=open('20171224_amplicon_genome.out','r')\n",
    "fahaha=Genome_blast.readlines()\n",
    "qt=0\n",
    "for z in range(len(fahaha)):\n",
    "    az=fahaha[z].split()\n",
    "    if float(az[2])!=100.000:\n",
    "        qt+=1\n",
    "        print(az[1], \" don't match with \", az[0], \" with \", az[4], \" mismatch(es) and \",az[5], \"gap(s).\")\n",
    "    \n",
    "    az=''\n",
    "if qt==0:\n",
    "    print(\"Every genomes and primers perfectly match.\") \n",
    "else:\n",
    "    print(\"There are \",qt,\" numbers of mismatch(es)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "expected an indented block (<ipython-input-1-af31756dde92>, line 14)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-1-af31756dde92>\"\u001b[0;36m, line \u001b[0;32m14\u001b[0m\n\u001b[0;31m    else:\u001b[0m\n\u001b[0m       ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m expected an indented block\n"
     ]
    }
   ],
   "source": [
    "#Q-5. Extract those amplicon sequences, and run the BLAST to all SARS-CoV-2 proteomes. Are they all perfectly matched?\n",
    "protein_blast=open('20171224_amplicon_protein.out','r')\n",
    "pepper=protein_blast.readlines()\n",
    "peanut=0\n",
    "for z in range(len(pepper)):\n",
    "    kitty=pepper[z].split()\n",
    "    if float(kitty[2])!=100.000:\n",
    "        peanut+=1\n",
    "        print(kitty[1], \" don't match with \", kitty[0], \" with \", kitty[4], \" mismatch(es) and \",kitty[5], \"gap(s).\")\n",
    "    \n",
    "    kitty=''\n",
    "if peanut==0:\n",
    "    #print(\"Every genomes and primers perfectly match.\") \n",
    "else:\n",
    "    #print(\"There are \",peanut,\" number of mismatch(es)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
