{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The function to read sequences from a FASTA file\n",
    "def read_fasta(tmp_filename):\n",
    "    rv = dict()\n",
    "    f = open(tmp_filename, 'r')\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            tmp_h = line.strip().lstrip('>')\n",
    "            rv[tmp_h] = ''\n",
    "        else:\n",
    "            rv[tmp_h] += line.strip().replace(' ', '')\n",
    "    f.close()\n",
    "    return rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'2019-nCoV_N1': {'F': 'GACCCCAAAATCAGCGAAAT', 'R': 'CAGATTCAACTGGCAGTAACCAGA', 'P': 'ACCCCGCATTACGTTTGGTGGACC'}, '2019-nCoV_N2': {'F': 'TTACAAACATTGGCCGCAAA', 'R': 'TTCTTCGGAATGTCGCGC', 'P': 'ACAATTTGCCCCCAGCGCTTCAG'}, '2019-nCoV_N3': {'F': 'GGGAGCCTTGAATACACCAAAA', 'R': 'CAATGCTGCAATCGTGCTACA', 'P': 'AYCACATTGGCACCCGCAATCCTG'}, 'RP': {'F': 'AGATTTGGACCTGCGAGCG', 'R': 'ACTTGTGGAGACAGCCGCTC', 'P': 'TTCTGACCTGAAGGCTCTGCGCG'}}\n",
      "there are 4 sets of primer\n"
     ]
    }
   ],
   "source": [
    "# Read primers\n",
    "filename_primers = '2019nCoV_primers.fa'\n",
    "primer_list = read_fasta(filename_primers)\n",
    "\n",
    "#make the new dictionaries, primer and types\n",
    "primer = dict()\n",
    "types = dict()\n",
    "\n",
    "#sort the primer according to their set\n",
    "for p_h, p_s in primer_list.items():   \n",
    "    \n",
    "    #apply revcomp function to reverse primer for genome sequnece comparison\n",
    "    if p_h[-1] == 'R':\n",
    "        p_s = revcomp(p_s)\n",
    "   \n",
    "    #primer.keys have primer set name / primer.values have types which is another dictionary\n",
    "    #types.keys have 3 type of primer, forward(F), reverse(R), prove(P) / types.values have primer sequences\n",
    "    if p_h[:-2] in primer.keys():  \n",
    "        primer[p_h[:-2]][p_h[-1]] = p_s       \n",
    "    else:\n",
    "        types[p_h[-1]] = p_s\n",
    "        primer[p_h[:-2]] = types\n",
    "        types = dict()\n",
    "        \n",
    "print(primer)\n",
    "print(\"there are\", len(primer),\"sets of primer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Q1: What is the length of amplicons generated by primers? Are they all same among the genomes?\n",
      "\n",
      " in 2019-nCoV_N1 , mean of amplicon lengths is 72.0\n",
      "** length of amplicons are all same among the genomes in 2019-nCoV_N1 primer sets\n",
      "\n",
      " in 2019-nCoV_N2 , mean of amplicon lengths is 67.0\n",
      "** length of amplicons are all same among the genomes in 2019-nCoV_N2 primer sets\n",
      "\n",
      " in 2019-nCoV_N3 , mean of amplicon lengths is 72.0\n",
      "** length of amplicons are all same among the genomes in 2019-nCoV_N3 primer sets\n",
      "\n",
      " no amplicons are detected by RP primer sets with these genome sets\n"
     ]
    }
   ],
   "source": [
    "# Q1. What is the length of amplicons generated by primers? Are they all same among the genomes?\n",
    "\n",
    "# Read genomes\n",
    "filename_genomes = '2019nCoV_genomes.2020_03_27.fa'\n",
    "genome_list = read_fasta(filename_genomes)\n",
    "\n",
    "A = dict()\n",
    "count = dict()\n",
    "SUM = 0\n",
    "\n",
    "# p_set: string, <2019-nCoV_N1,2019-nCoV_N2,2019-nCoV_N3, RP>\n",
    "# p_type: dictionary, <F:seqeunces, R:sequences. P:sequences>\n",
    "print(\"Q1: What is the length of amplicons generated by primers? Are they all same among the genomes?\")\n",
    "for p_set, p_type in primer.items():\n",
    "    for g_h, g_seq in genome_list.items():\n",
    "        \n",
    "        #p_type[\"F\"]: forward primer sequence, p_type[\"R\"]: reverse primer sequnece\n",
    "        if p_type[\"F\"] in g_seq and p_type[\"R\"] in g_seq:\n",
    "            # we have to add the reverse primer sequence length, because amplicon include both forward and reverse primer\n",
    "            amplicon_len = abs(g_seq.find(p_type[\"R\"])-g_seq.find(p_type[\"F\"]))+len(p_type[\"R\"])\n",
    "            A[g_h] = amplicon_len            \n",
    "    \n",
    "    #count the overlapping amplicon lengths and \n",
    "    for c in A.values():\n",
    "        SUM = SUM + c\n",
    "        if c not in count.keys():\n",
    "            count[c] = 1\n",
    "        else:\n",
    "            count[c] += 1\n",
    "    \n",
    "    #print the results\n",
    "    if len(count) == 1:\n",
    "        print(\"\\n in\", p_set,\", mean of amplicon lengths is\", SUM/len(A))\n",
    "        print(\"** length of amplicons are all same among the genomes in\", p_set,\"primer sets\")\n",
    "    elif len(count) == 0:\n",
    "        print(\"\\n no amplicons are detected by\",p_set, \"primer sets with these genome sets\")\n",
    "    else:\n",
    "        print(\"in\", p_set,\", mean of amplicon lengths is\", SUM/len(A))\n",
    "        print(\"\\n amplicon size are various among the genomes\")\n",
    "    \n",
    "    #reset the SUM, A, count when start with new primer sets\n",
    "    SUM = 0\n",
    "    A = dict() \n",
    "    count = dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Q2. Can these primers detect all SARS-CoV-2 genomes?\n",
      "\n",
      " 2019-nCoV_N1 primer can detect all SARS-CoV-2 genomes\n",
      "\n",
      " 2019-nCoV_N2 primer can detect all SARS-CoV-2 genomes\n",
      "\n",
      " 2019-nCoV_N3 primer can detect all SARS-CoV-2 genomes\n",
      "\n",
      " RP primer can't detect any SARS-CoV-2 genomes\n"
     ]
    }
   ],
   "source": [
    "# Q2. Can these primers detect all SARS-CoV-2 genomes?\n",
    "# '../data/2019nCoV_genomes.2020_02_03.fa' and '../data/2019nCoV_genomes.2020_03_27.fa'\n",
    "\n",
    "# Read genomes of SARS-CoV-2 genomes\n",
    "filename_genomes = '2019nCoV_genomes.2020_02_03.fa'\n",
    "genome_list_old = read_fasta(filename_genomes)\n",
    "\n",
    "# add 2019_nCoV_genomes.2020_02_03.fa genome list to my genome list dictionary\n",
    "genome_list.update(genome_list_old)\n",
    "\n",
    "count = 0\n",
    "# p_set: string, <2019-nCoV_N1,2019-nCoV_N2,2019-nCoV_N3, RP>\n",
    "# p_type: dictionary, <F:seqeunces, R:sequences. P:sequences>\n",
    "print(\"Q2. Can these primers detect all SARS-CoV-2 genomes?\")\n",
    "for p_set, p_type in primer.items():\n",
    "    for g_h, g_seq in genome_list.items():\n",
    "        \n",
    "        #p_type[\"F\"]: forward primer sequence, p_type[\"R\"]: reverse primer sequnece\n",
    "        if p_type[\"F\"] in g_seq and p_type[\"R\"] in g_seq:\n",
    "            count += 1            \n",
    "    \n",
    "    #print the results\n",
    "    if len(genome_list) == count:\n",
    "        print(\"\\n\",p_set,\"primer can detect all SARS-CoV-2 genomes\")\n",
    "    elif count == 0:\n",
    "        print(\"\\n\",p_set,\"primer can't detect any SARS-CoV-2 genomes\")\n",
    "    else:\n",
    "        print(\"\\n\",p_set,\"primer can't detect all SARS-CoV-2 genomes, but can detect\",count,\"types of SARS-CoV-2 genomes\")\n",
    "        \n",
    "    \n",
    "    #reset count when start with new primer sets\n",
    "    count = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Q3. Can these primers detect MERS genomes? How about SARS genomes?\n",
      "\n",
      " 2019-nCoV_N1 primer set can't detect both MERS and SARS genome\n",
      "\n",
      " 2019-nCoV_N2 primer set can't detect both MERS and SARS genome\n",
      "\n",
      " 2019-nCoV_N3 primer set can't detect both MERS and SARS genome\n",
      "\n",
      " RP primer set can't detect both MERS and SARS genome\n"
     ]
    }
   ],
   "source": [
    "# Q3. Can these primers detect MERS genomes? How about SARS genomes? (you can find those genomes under '../data' directory).\n",
    "\n",
    "# Read genomes\n",
    "MERS = 'MERS_genomes.2020_02_03.fa'\n",
    "SARS = 'SARS_genomes.2020_02_03.fa'\n",
    "MERS_list = read_fasta(MERS)\n",
    "SARS_list = read_fasta(SARS)\n",
    "S_count = 0\n",
    "M_count = 0\n",
    "\n",
    "# p_set: string, <2019-nCoV_N1,2019-nCoV_N2,2019-nCoV_N3, RP>\n",
    "# p_type: dictionary, <F:seqeunces, R:sequences. P:sequences>\n",
    "print(\"Q3. Can these primers detect MERS genomes? How about SARS genomes?\")\n",
    "for p_set, p_type in primer.items():\n",
    "    \n",
    "    for g_h, g_seq in MERS_list.items():\n",
    "        \n",
    "        #p_type[\"F\"]: forward primer sequence, p_type[\"R\"]: reverse primer sequnece\n",
    "        if p_type[\"F\"] in g_seq and p_type[\"R\"] in g_seq:\n",
    "            M_count += 1            \n",
    "    \n",
    "    for g_h, g_seq in SARS_list.items():\n",
    "        \n",
    "        #p_type[\"F\"]: forward primer sequence, p_type[\"R\"]: reverse primer sequnece\n",
    "        if p_type[\"F\"] in g_seq and p_type[\"R\"] in g_seq:\n",
    "            S_count += 1            \n",
    "\n",
    "    #print the results\n",
    "    if M_count == 0 and S_count == 0:\n",
    "        print(\"\\n\",p_set,\"primer set can't detect both MERS and SARS genome\")  \n",
    "    elif M_count != 0:\n",
    "        print(\"\\n\",p_set, \"primer set can detect MERS genome\")\n",
    "    elif S_count != 0:\n",
    "        print(\"\\n\",p_set, \"primer set can detect SARS genome\")\n",
    "    else:\n",
    "        print(\"\\n\",p_set, \"primer set can detect both MERS and SARS genomes\")\n",
    "    \n",
    "    #reset S_count and M_count when start with new primer set\n",
    "    S_count = 0\n",
    "    M_count = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I copied Hyerinkim's work tried to understand whats going on\n",
    "# this work doesnt belong to me\n",
    "#sorry fot that(("
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
