{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task_02. In Silico PCR for Covid-19 diagnosis.\n",
    "CDC(Center for Disease Control and Prevention) released the information about PCR primers/probes to detect Covid-19. (See [this page](https://www.cdc.gov/coronavirus/2019-ncov/lab/rt-pcr-panel-primer-probes.html) for more information). We are curious how these primers/probes work.\n",
    "\n",
    "If you need more info about \"PCR-based diagnosis\", see [this video](https://www.youtube.com/watch?v=fkUDu042xic).\n",
    "\n",
    "## Data files\n",
    "<ul>\n",
    "    <li> The genome of Covid-19: '../data/2019nCoV_genomes.2020_02_03.fa'\n",
    "    <li> The primers for Covid-19 detection: '../data/2019nCoV_primers.fa'\n",
    "</ul>\n",
    "    \n",
    "## Procedures\n",
    "<ol>\n",
    "<li> Read 2019nCoV primers from a FASTA file (**see below**).\n",
    "<li> Read 2019nCoV genomes from a FASTA file (**see below**).\n",
    "<li> Find the position of primers (F, R) on each genome sequence.\n",
    "<li> Calculate the length of PCR amplicons for Covid-19 diagnostics.\n",
    "</ol>\n",
    "\n",
    "## Questions\n",
    "<ol>\n",
    "    <li> What is the length of amplicons generated by N1/N2/N3 primers? Any variation among the genomes?\n",
    "    <li> What is the sequence of amplicons?\n",
    "    <li> Can these primers detect all Covid-19 genomes?\n",
    "    <li> Can these primers detect MERS genomes? How about SARS genomes? (you can find those genomes under '../data' directory).\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The function to get the reverse complementary sequences\n",
    "rs = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}\n",
    "def comp(seq):    \n",
    "    return ''.join([rs[x] for x in seq[::-1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The function to read sequences from a FASTA file\n",
    "def read_fasta(tmp):\n",
    "    r = dict()\n",
    "    fi = open(tmp, 'r')\n",
    "    for line in fi:\n",
    "        if line.startswith('>'):\n",
    "            h_g = line.strip().lstrip('>')\n",
    "            r[h_g] = ''\n",
    "        else:\n",
    "            r[h_g] += line.strip().replace(' ', '')\n",
    "    fi.close()\n",
    "    return r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read primers\n",
    "\n",
    "primers_set = '../data/2019nCoV_primers.fa'\n",
    "primer = read_fasta(primers_set)\n",
    "primer['nCoV_IP2-F'] = 'ATGAGCTTAGTCCTGTTG'\n",
    "primer['nCoV_IP2-R'] = 'CTCCCTTTGTTGTGTTGT'\n",
    "primer['nCoV_IP2-P'] = 'AGATGTCTTGTGCTGCCGGTA'\n",
    "\n",
    "F_P=[]\n",
    "R_C=[]\n",
    "P_P=[]\n",
    "\n",
    "for p_h, p_seq in primer.items():\n",
    "    if '-F' in p_h:\n",
    "        F_P.append(p_seq)\n",
    "    elif '-R' in p_h:\n",
    "        C=comp(p_seq)\n",
    "        R_C.append(C)\n",
    "    else:\n",
    "        P_P.append(p_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read genomes\n",
    "SARS_CoV2_genomes = '../data/2019nCoV_genomes.2020_02_03.fa'\n",
    "genome_COVID = read_fasta(SARS_CoV2_genomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Counting index of Forward primer and Reverse primer\n",
    "def count_idx(p, g):\n",
    "    n=len(g)\n",
    "    leng=len(p)\n",
    "    for i in range(0, n-leng+1):\n",
    "        if g[i : i+leng]==p:\n",
    "            return i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Return amplicon seq\n",
    "def detect_amplicon(fp, rc, g):\n",
    "    amplicon=''\n",
    "    if fp in g and rc in g:\n",
    "        a=count_idx(fp, g)\n",
    "        b=count_idx(rc, g)\n",
    "        j=len(fp)\n",
    "        o=len(rc)\n",
    "        if a + j < b:\n",
    "            amplicon=g[a : b + o]\n",
    "    return amplicon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2019-nCoV_N3 : 72\n",
      "2019-nCoV_N2 : 67\n",
      "nCoV_IP2 : 108\n",
      "2019-nCoV_N1 : 72\n",
      "RP : 0\n"
     ]
    }
   ],
   "source": [
    "# HW 2-1-1. length of amplicons by primers\n",
    "g_seq = list(genome_COVID.values())\n",
    "p_N = [p_h[:-2] for p_h in primer.keys()]\n",
    "amplicon_s = {val : [] for val in set(p_N)}\n",
    "\n",
    "for g in g_seq:\n",
    "    for lie in range(len(F_P)):\n",
    "        fp = F_P[lie]\n",
    "        rc = R_C[lie]\n",
    "        le=lie*3\n",
    "        amplicon = detect_amplicon(fp, rc, g)\n",
    "        amplicon_s[p_N[le]].append(amplicon)\n",
    "        \n",
    "for val in amplicon_s.keys():\n",
    "    print(val, \": {}\".format(len(amplicon_s[val][0])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2019-nCoV_N3 : True\n",
      "2019-nCoV_N2 : True\n",
      "nCoV_IP2 : True\n",
      "2019-nCoV_N1 : True\n"
     ]
    }
   ],
   "source": [
    "# HW 2-1-2. Are they all same among the genomes?\n",
    "for val in amplicon_s.keys():\n",
    "    if len(amplicon_s[val][0])!=0:\n",
    "        if len(set(amplicon_s[val])) == 1 and set(amplicon_s[val]) != {''}:\n",
    "            print(val, \": True\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2019-nCoV_N3 primer can detect 15 cases in 15 total sample genomes.\n",
      "2019-nCoV_N2 primer can detect 15 cases in 15 total sample genomes.\n",
      "nCoV_IP2 primer can detect 15 cases in 15 total sample genomes.\n",
      "2019-nCoV_N1 primer can detect 15 cases in 15 total sample genomes.\n",
      "RP primer cannot detect 15 cases in 15 total sample genomes.\n"
     ]
    }
   ],
   "source": [
    "#HW 2-2 SARS-CoV2 case\n",
    "for val in amplicon_s.keys():\n",
    "    if '' in list(amplicon_s[val]):\n",
    "        print(val, \"primer cannot detect {} cases in {} total sample genomes.\".format(amplicon_s[val].count(''), len(amplicon_s[val])))\n",
    "    else:\n",
    "        print(val, \"primer can detect {} cases in {} total sample genomes.\".format(len(amplicon_s[val]), len(amplicon_s[val])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2019-nCoV_N3 cannot detect MERS.\n",
      "2019-nCoV_N2 cannot detect MERS.\n",
      "nCoV_IP2 cannot detect MERS.\n",
      "2019-nCoV_N1 cannot detect MERS.\n",
      "RP cannot detect MERS.\n"
     ]
    }
   ],
   "source": [
    "#HW 2-3-1 MERS case\n",
    "MERS_genomes = '../data/MERS_genomes.2020_02_03.fa'\n",
    "genome_MERS = read_fasta(MERS_genomes)\n",
    "M_seq = list(genome_MERS.values())\n",
    "M_amplicon_s = {val : [] for val in set(p_N)}\n",
    "\n",
    "for gM in M_seq:\n",
    "    for lie in range(len(F_P)):\n",
    "        fp = F_P[lie]\n",
    "        rc = R_C[lie]\n",
    "        le=lie*3\n",
    "        amplicon = detect_amplicon(fp, rc, gM)\n",
    "        M_amplicon_s[p_N[le]].append(amplicon)\n",
    "        \n",
    "\n",
    "for val in M_amplicon_s.keys():\n",
    "    if set(M_amplicon_s[val]) == {''}:\n",
    "        print(val, \"cannot detect MERS.\")\n",
    "    else:\n",
    "        if '' in list(M_amplicon_s[val]):\n",
    "            print(val, \"primer cannot detect {} cases in {} total sample genomes.\".format(M_amplicon_s[val].count(''), len(M_amplicon_s[val])))\n",
    "        else:\n",
    "            print(val, \"primer can detect {} cases in {} total sample genomes.\".format(len(M_amplicon_s[val]), len(M_amplicon_s[val])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2019-nCoV_N3 cannot detect SARS.\n",
      "2019-nCoV_N2 cannot detect SARS.\n",
      "nCoV_IP2 cannot detect SARS.\n",
      "2019-nCoV_N1 cannot detect SARS.\n",
      "RP cannot detect SARS.\n"
     ]
    }
   ],
   "source": [
    "#HW 2-3-2 SARS case\n",
    "SARS_genomes = '../data/SARS_genomes.2020_02_03.fa'\n",
    "genome_SARS = read_fasta(SARS_genomes)\n",
    "S_seq = list(genome_SARS.values())\n",
    "S_amplicon_s = {val : [] for val in set(p_N)}\n",
    "\n",
    "for gS in S_seq:\n",
    "    for lie in range(len(F_P)):\n",
    "        fp = F_P[lie]\n",
    "        rc = R_C[lie]\n",
    "        le=lie*3\n",
    "        amplicon = detect_amplicon(fp, rc, gS)\n",
    "        S_amplicon_s[p_N[le]].append(amplicon)\n",
    "        \n",
    "\n",
    "for val in S_amplicon_s.keys():\n",
    "    if set(S_amplicon_s[val]) == {''}:\n",
    "        print(val, \"cannot detect SARS.\")\n",
    "    else:\n",
    "        if '' in list(S_amplicon_s[val]):\n",
    "            print(val, \"primer cannot detect {} cases in {} total sample genomes.\".format(S_amplicon_s[val].count(''), len(S_amplicon_s[val])))\n",
    "        else:\n",
    "            print(val, \"primer can detect {} cases in {} total sample genomes.\".format(len(S_amplicon_s[val]), len(S_amplicon_s[val])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">2019-nCoV_N3->Amplicons\n",
      "GGGAGCCTTGAATACACCAAAAGATCACATTGGCACCCGCAATCCTGCTAACAATGCTGCAATCGTGCTACA\n",
      ">2019-nCoV_N2->Amplicons\n",
      "TTACAAACATTGGCCGCAAATTGCACAATTTGCCCCCAGCGCTTCAGCGTTCTTCGGAATGTCGCGC\n",
      ">nCoV_IP2->Amplicons\n",
      "ATGAGCTTAGTCCTGTTGCACTACGACAGATGTCTTGTGCTGCCGGTACTACACAAACTGCTTGCACTGATGACAATGCGTTAGCTTACTACAACACAACAAAGGGAG\n",
      ">2019-nCoV_N1->Amplicons\n",
      "GACCCCAAAATCAGCGAAATGCACCCCGCATTACGTTTGGTGGACCCTCAGATTCAACTGGCAGTAACCAGA\n"
     ]
    }
   ],
   "source": [
    "#HW 2-4-1 Extract amplicons\n",
    "def w_amplicon_s(fi, gc):\n",
    "    f = open(fi, 'w')\n",
    "    for g in list(gc.keys()):\n",
    "        if len(gc[g]) == 1:\n",
    "            f.write(\">{}\\n\".format(g))\n",
    "            f.write(\"{}\\n\".format(gc[g][0]))\n",
    "        else:\n",
    "            for i in range(len(gc[g])):\n",
    "                f.write(\">{} #{}\\n\".format(g, i+1))\n",
    "                f.write(\"\\n{}\".format(gc[g][i]))\n",
    "    f.close()\n",
    "    \n",
    "amplicon_s_no = {\"{}->Amplicons\".format(val) : [] for val in set(p_N)}\n",
    "for val in list(amplicon_s.keys()):\n",
    "    amplicon_s_no[\"{}->Amplicons\".format(val)] = list(set(amplicon_s[val]))\n",
    "                    \n",
    "for val in list(amplicon_s.keys()):\n",
    "        if '' in amplicon_s_no[\"{}->Amplicons\".format(val)]:\n",
    "            amplicon_s_no[\"{}->Amplicons\".format(val)].remove('')\n",
    "            \n",
    "w_amplicon_s(\"20191116_SeonjunPark_HW2_amp.fa\", amplicon_s_no)\n",
    "f = open(\"20191116_SeonjunPark_HW2_amp.fa\", \"r\")\n",
    "for line in f.readlines():\n",
    "    print(line, end='')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#HW 2-4-2 run the BLAST to all SARS-CoV-2 genomes.\n",
    "f=open('20191116_SeonjunPark_HW2_blastn.out', 'r')\n",
    "\n",
    "bool=0\n",
    "\n",
    "for line in f.readlines():\n",
    "    blast_id = str(line).split()\n",
    "    if float(blast_id[2]) < 99.5:\n",
    "        print(\"{} and {} have {} mismatches and {} gaps\".format(blast_id[0], blast_id[1], blast_id[4], blast_id[5]))\n",
    "        bool += 1\n",
    "if bool==0:\n",
    "    print(\"Everything matches perfectly.\", '\\n')\n",
    "\n",
    "f.close()    \n",
    "f = open('20191116_SeonjunPark_HW2_blastn.out', 'r')\n",
    "for line in f.readlines():\n",
    "    print(line, end='')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '20191116_SeonjunPark_HW2_blastn.out'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-92a2307ad340>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#HW 2-4-2 run the BLAST to all SARS-CoV-2 genomes.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mf\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'20191116_SeonjunPark_HW2_blastn.out'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mbool\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mline\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadlines\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mblast_id\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '20191116_SeonjunPark_HW2_blastn.out'"
     ]
    }
   ],
   "source": [
    "#HW 2-5 run the BLAST to all SARS-CoV-2 genomes.\n",
    "f = open('20191116_SeonjunPark_HW2_blastx.out', 'r')\n",
    "\n",
    "bool= 0\n",
    "\n",
    "for line in f:\n",
    "    blastx_id = str(line).split()\n",
    "    if line.split()[4] != '0' or line.split()[5] != '0':\n",
    "        print('{} and {} have {} mismatches and {} gaps'\n",
    "              % (blastx_id[0], blastx_id[1], blastx_id[4], blastx_id[5]))\n",
    "        bool += 1\n",
    "if bool==0:\n",
    "    print('there is no mismatch or gap')        \n",
    "f.close()\n",
    "\n",
    "f = open('20191116_SeonjunPark_HW2_blastx.out', 'r')\n",
    "for line in f.readlines():\n",
    "    print(line, end='')\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Although I continue to try to blastn and blastx to do my HW 2-4-2 and 2-5, code errors continue to appear so I only uploaded my code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
