{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homework 01. Hojune Lee (Student ID - 20161333)\n",
    "CDC(Center for Disease Control and Prevention) released the information about PCR primers/probes to detect Covid-19. (See [this page](https://www.cdc.gov/coronavirus/2019-ncov/lab/rt-pcr-panel-primer-probes.html) for more information). We are curious how these primers/probes work.\n",
    "\n",
    "If you need more info about \"PCR-based diagnosis\", see [this video](https://www.youtube.com/watch?v=fkUDu042xic).\n",
    "\n",
    "## Data files\n",
    "<ul>\n",
    "    <li> The genome of Covid-19: '../data/2019nCoV_genomes.2020_02_03.fa'\n",
    "    <li> The primers for Covid-19 detection: '../data/2019nCoV_primers.fa'\n",
    "</ul>\n",
    "    \n",
    "## Procedures\n",
    "<ol>\n",
    "<li> Read 2019nCoV primers from a FASTA file (**see below**).\n",
    "<li> Read 2019nCoV genomes from a FASTA file (**see below**).\n",
    "<li> Find the position of primers (F, R) on each genome sequence.\n",
    "<li> Calculate the length of PCR amplicons for Covid-19 diagnostics.\n",
    "</ol>\n",
    "\n",
    "## Questions\n",
    "<ol>\n",
    "    <li> What is the length of amplicons generated by N1/N2/N3 primers? Any variation among the genomes?\n",
    "    <li> What is the sequence of amplicons?\n",
    "    <li> Can these primers detect all Covid-19 genomes?\n",
    "    <li> Can these primers detect MERS genomes? How about SARS genomes? (you can find those genomes under '../data' directory).\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The function to read sequences from a FASTA file\n",
    "def read_fasta(tmp_filename):\n",
    "    rv = dict()\n",
    "    f = open(tmp_filename, 'r')\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            tmp_h = line.strip().lstrip('>')\n",
    "            rv[tmp_h] = ''\n",
    "        else:\n",
    "            rv[tmp_h] += line.strip().replace(' ', '')\n",
    "    f.close()\n",
    "    return rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENA|MN988668|MN988668.1 TTAAAGGTTTATACCTTCCCAGGTA\n",
      "ENA|MN988669|MN988669.1 TTAAAGGTTTATACCTTCCCAGGTA\n",
      "ENA|MN938384|MN938384.1 CAACCAACTTTCGATCTCTTGTAGA\n",
      "ENA|MN975262|MN975262.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN996527|MN996527.1 AACCAACTTTCGATCTCTTGTAGAT\n",
      "ENA|MN996528|MN996528.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN996529|MN996529.1 TACCTTCCCAGGTAACAAACCAACC\n",
      "ENA|MN996530|MN996530.1 CCTTCCCAGGTAACAAACCAACCAA\n",
      "ENA|MN996531|MN996531.1 ACCTTCCCAGGTAACAAACCAACCA\n",
      "ENA|MN908947|MN908947.3 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN985325|MN985325.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN988713|MN988713.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN994467|MN994467.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN994468|MN994468.1 ATTAAAGGTTTATACCTTCCCAGGT\n",
      "ENA|MN997409|MN997409.1 ATTAAAGGTTTATACCTTCCCAGGT\n"
     ]
    }
   ],
   "source": [
    "# Read genomes\n",
    "filename_genomes = '../data/2019nCoV_genomes.2020_02_03.fa'\n",
    "genome_list = read_fasta(filename_genomes)\n",
    "for g_h, g_seq in genome_list.items():\n",
    "    print(g_h.split()[0], g_seq[:25])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sequences: 15\n"
     ]
    }
   ],
   "source": [
    "#HW 1-1\n",
    "#Number of sequences\n",
    "\n",
    "num_seq = len(genome_list)\n",
    "print(\"Number of sequences:\", num_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of sequences: [29881, 29881, 29838, 29891, 29825, 29891, 29852, 29854, 29857, 29903, 29882, 29882, 29882, 29883, 29882]\n"
     ]
    }
   ],
   "source": [
    "#HW 1-2\n",
    "#Length of sequences\n",
    "length_of_sequences = []\n",
    "for g_seq in genome_list.values():\n",
    "    length_of_sequences.append(len(g_seq))\n",
    "print(\"Length of sequences:\", length_of_sequences)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENA|MN988668|MN988668.1', 'ENA|MN988669|MN988669.1'] are identical with each other.\n",
      "number of identical sequences: 1\n"
     ]
    }
   ],
   "source": [
    "#HW 1-3\n",
    "#Number of identical sequences + their names\n",
    "\n",
    "identical_sequences = []\n",
    "name_of_sequences = list(genome_list.keys())\n",
    "sequences_to_check = list(genome_list.values())\n",
    "\n",
    "#erase a sequence from the list and see if the list has the same ones. \n",
    "while(len(sequences_to_check) > 1):\n",
    "    current_sequence = sequences_to_check.pop(0)\n",
    "    if current_sequence not in sequences_to_check:\n",
    "        name_of_sequences.pop(0)\n",
    "    else:\n",
    "        templist = []\n",
    "        templist.append(name_of_sequences.pop(0).split()[0])\n",
    "        while current_sequence in sequences_to_check:\n",
    "            idx = sequences_to_check.index(current_sequence)\n",
    "            sequences_to_check.pop(idx)\n",
    "            templist.append(name_of_sequences.pop(idx).split()[0])\n",
    "        identical_sequences.append(templist)\n",
    "        \n",
    "\n",
    "for sequences in identical_sequences:\n",
    "    print(sequences, \"are identical with each other.\")\n",
    "print(\"number of identical sequences: {}\".format(len(identical_sequences)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read primers\n",
    "\n",
    "filename_primers = '../data/2019nCoV_primers.fa'\n",
    "primer_list = read_fasta(filename_primers)\n",
    "for p_h, p_seq in primer_list.items():\n",
    "    print(p_h, p_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A function that consider IUPAC Nucleotide code when comparing codes\n",
    "# See whether a probe which has codes such as R, Y and S is a part of a genome which has only A,T,G and C for its codes.\n",
    "\n",
    "def IsProbe1(probe, genome):\n",
    "\tpossible_probes = [probe]\n",
    "\tfor idx in range(len(probe)):\n",
    "\t\tif probe[idx] == 'R':\n",
    "\t\t\tpossible_probes = (\" \".join(possible_probes).replace(probe[idx], 'A', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'G', 1)).split(\" \")\n",
    "\t\telif probe[idx] == 'Y':\n",
    "\t\t\tpossible_probes = (\" \".join(possible_probes).replace(probe[idx], 'C', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'T', 1)).split(\" \")\n",
    "\t\telif probe[idx] == 'S':\n",
    "\t\t\tpossible_probes = (\" \".join(possible_probes).replace(probe[idx], 'G', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'C', 1)).split(\" \")\n",
    "\t\telif probe[idx] == 'W':\n",
    "\t\t\tpossible_probes = (\" \".join(possible_probes).replace(probe[idx], 'A', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'T', 1)).split(\" \")\n",
    "\t\telif probe[idx] == 'K':\n",
    "\t\t\tpossible_probes = (\" \".join(possible_probes).replace(probe[idx], 'G', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'T', 1)).split(\" \")\n",
    "\t\telif probe[idx] == 'M':\n",
    "\t\t\tpossible_probes = (\" \".join(possible_probes).replace(probe[idx], 'A', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'C', 1)).split(\" \")\n",
    "\t\telif probe[idx] == 'B':\n",
    "\t\t\tpossible_probes = (\" \".join(possible_probes).replace(probe[idx], 'C', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'G', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'T', 1)).split(\" \")\n",
    "\t\telif probe[idx] == 'D':\n",
    "\t\t\tpossible_probes = (\" \".join(possible_probes).replace(probe[idx], 'A', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'G', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'T', 1)).split(\" \")\n",
    "\t\telif probe[idx] == 'H':\n",
    "\t\t\tpossible_probes = (\" \".join(possible_probes).replace(probe[idx], 'A', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'C', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'T', 1)).split(\" \")\n",
    "\t\telif probe[idx] == 'V':\n",
    "\t\t\tpossible_probes = (\" \".join(possible_probes).replace(probe[idx], 'A', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'C', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'G', 1)).split(\" \")\n",
    "\t\telif probe[idx] == 'N':\n",
    "\t\t\tpossible_probes = (\" \".join(possible_probes).replace(probe[idx], 'A', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'G', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'T', 1) + \" \" + (\" \".join(possible_probes)).replace(probe[idx], 'C', 1)).split(\" \")\n",
    "\tfor possible_probe in possible_probes:\n",
    "\t\tif possible_probe in genome:\n",
    "\t\t\treturn True\n",
    "\t\telse:\n",
    "\t\t\tcontinue\n",
    "\treturn False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The fuction to get an amplicon from Primers F, R, P and a genome to be PCRed\n",
    "# If they doesn't matches, return ''.\n",
    "def primerFRP_to_amplicon(F_primer, R_primer, P_primer, genome):\n",
    "    ap = ''\n",
    "    if F_primer not in genome:\n",
    "        return ap\n",
    "    if revcomp(R_primer) not in genome:\n",
    "        return ap\n",
    "    if F_primer in genome and revcomp(R_primer) in genome:\n",
    "        print(genome[:50])\n",
    "        ap = genome[genome.index(F_primer):genome.index(revcomp(R_primer))] + revcomp(R_primer)\n",
    "        return ap\n",
    "    return ''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
