{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homework 03. Jeun Hyun (Student ID -20191328)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The function to read sequences from a FASTA file\n",
    "def read_fasta(tmp_filename):\n",
    "    rv = dict()\n",
    "    f = open(tmp_filename, 'r')\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            tmp_h = line.strip().lstrip('>')\n",
    "            rv[tmp_h] = ''\n",
    "        else:\n",
    "            rv[tmp_h] += line.strip().replace(' ', '')\n",
    "    f.close()\n",
    "    return rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_009858895.2 genomic sequence is ATTAAAGGTTTATACCTTCCCAGGT ... AAAAAAAAAAAAAAAAAAAAAAAAA\n"
     ]
    }
   ],
   "source": [
    "# Read GCF_009858895.2 genomic info.\n",
    "filename_genomes = '../data/GCF_009858895.2_ASM985889v3_genomic.fna'\n",
    "genome_list = read_fasta(filename_genomes)\n",
    "genome_seq = list(genome_list.values())[0]\n",
    "\n",
    "print('GCF_009858895.2 genomic sequence is', genome_seq[:25], '...', genome_seq[-25:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N-protein cDNA sequence is ATGTCTGATAATGGACCCCAAAATC ... CAGTGCTGACTCAACTCAGGCCTAA\n"
     ]
    }
   ],
   "source": [
    "# Make N-protein cDNA sequence.\n",
    "f = open('../data/GCF_009858895.2_ASM985889v3_genomic.gff', 'r')\n",
    "\n",
    "for line in f:\n",
    "    try:\n",
    "        if 'Name=N' in line.split()[8]:\n",
    "            N_genomic = genome_seq[int(line.split()[3])-1 : int(line.split()[4])]\n",
    "            print('N-protein cDNA sequence is', N_genomic[:25], '...', N_genomic[-25:])\n",
    "    except:\n",
    "        continue\n",
    "        \n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N-protein sequence is MSDNGPQNQRNAPRITFGGPSDSTG ... LPAADLDDFSKQLQQSMSSADSTQA\n"
     ]
    }
   ],
   "source": [
    "# Translate it to protein sequence. \n",
    "f = open(\"N_genomic.fa\", 'w')\n",
    "data = \"> N_genomic\\n%s\\n\" % N_genomic\n",
    "f.write(data)\n",
    "f.close()\n",
    "\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ transeq --sequence N_genomic.fa --outseq N_trans.fa\n",
    "\n",
    "N_protein = list(read_fasta('N_trans.fa').values())[0][:-1] # exclude not-designated codon - stop codon\n",
    "print('N-protein sequence is', N_protein[:25], '...', N_protein[-25:]) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The translated sequence and the sequence in data of N_protein(nucleocapsid phosphoprotein) are not identical\n"
     ]
    }
   ],
   "source": [
    "# Q1. Compare it to GCF_009858895.2_ASM985889v3_protein.faa\n",
    "f = open('../data/GCF_009858895.2_ASM985889v3_protein.faa', 'r')\n",
    "protein_list = read_fasta('../data/GCF_009858895.2_ASM985889v3_protein.faa')\n",
    "protein_h = list(protein_list.keys())\n",
    "protein_seq = list(protein_list.values())\n",
    "\n",
    "detect = 0\n",
    "for i in range(len(protein_h)):\n",
    "    if 'nucleocapsid phosphoprotein' in protein_h[i]:\n",
    "        detect = 1\n",
    "        if N_protein[:-1] == protein_seq[i]:\n",
    "            print('The translated sequence and the sequence in data of N_protein(nucleocapsid phosphoprotein) are identical')\n",
    "        else:\n",
    "            print('The translated sequence and the sequence in data of N_protein(nucleocapsid phosphoprotein) are not identical')\n",
    "if detect == 0:\n",
    "    print('Cannot find N_protein(nucleocapsid phosphoprotein) sequence in data')\n",
    "        \n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Based on this info, extract N proteins cDNA/proteins from other SARS-CoV2 genomes (03_27).\n",
    "\n",
    "# in cDNA\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ makeblastdb -dbtype nucl -in ../data/2019nCoV_genomes.2020_03_27.fa -out 2019nCoV_genomes.2020_03_27\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ blastn -db 2019nCoV_genomes.2020_03_27 -query N_genomic.fa -outfmt 6 -out 2019nCoV_genomes.2020_03_27_blastn.out\n",
    "\n",
    "f = open('2019nCoV_genomes.2020_03_27_blastn.out', 'r')\n",
    "\n",
    "genome_list = read_fasta('../data/2019nCoV_genomes.2020_03_27.fa')\n",
    "for line in f:\n",
    "    for genome in list(genome_list.keys()):\n",
    "        if line.split()[1] in genome:\n",
    "            print(end='')\n",
    "#            print(genome_list[genome][int(line.split()[8])-1:int(line.split()[9])], 'from', genome.split()[0])\n",
    "    \n",
    "f.close()\n",
    "\n",
    "\n",
    "# in protein\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ transeq --sequence ../data/2019nCoV_genomes.2020_03_27.fa --outseq ../data/2019nCoV_genomes.2020_03_27_trans6.fa -frame 6\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ makeblastdb -dbtype prot -in ../data/2019nCoV_genomes.2020_03_27_trans6.fa -out ../data/2019nCoV_genomes.2020_03_27_trans6\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ blastp -db ../data/2019nCoV_genomes.2020_03_27_trans6 -query N_trans.fa -outfmt 6 -out 2019nCoV_genomes.2020_03_27_trans6_blastp.out\n",
    "\n",
    "f = open('2019nCoV_genomes.2020_03_27_trans6_blastp.out', 'r')\n",
    "\n",
    "genome_list = read_fasta('../data/2019nCoV_genomes.2020_03_27.fa')\n",
    "for line in f:\n",
    "    for genome in list(genome_list.keys()):\n",
    "        if line.split()[1] in genome:\n",
    "            print(end='')\n",
    "#            print(genome_list[genome][int(line.split()[8])-1:int(line.split()[9])], 'from', genome.split()[0])\n",
    "    \n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "In terms of cDNA\n",
      "MT049951 MT066176 MT066175 MT072688 MT093571 MT093631 MT123292 LC528232 LC528233 MT126808 MT135044 MT135042 MT135041 MT152824 MT050493 MT012098 MT066156 MT121215 MT163718 MT163717 MT163716 MT163719 MT192773 MT192759 MT192765 MT192772 MT240479 MT246490 MT246489 MT246488 MT246487 MT246486 MT246484 MT246482 MT246481 MT246480 MT246479 MT246478 MT246477 MT246476 MT246475 MT246474 MT246473 MT246472 MT246471 MT246469 MT246468 MT246467 MT246466 MT246464 MT246462 MT246461 MT246460 MT246459 MT246458 MT246457 MT246455 MT246453 MT246452 MT246450 MT246449 are the most similar ones by 100.000000 percent identity\n",
      "MT233522 MT246470 MT246451 are the most different ones by 99.762000 percent identity\n",
      "\n",
      "\n",
      "In terms of protein\n",
      "MT049951_2 MT066176_2 MT066175_2 MT072688_2 MT093571_2 MT093631_1 MT123293_2 MT123291_2 MT123292_2 LC528232_2 LC528233_1 MT126808_2 MT135044_2 MT135042_2 MT135041_2 MT152824_3 MT050493_3 MT012098_1 MT066156_2 MT121215_2 MT163718_2 MT163717_2 MT163716_2 MT163719_2 MT192773_1 MT192759_2 MT192765_1 MT192772_2 MT240479_2 MT246490_3 MT246489_3 MT246488_2 MT246487_1 MT246486_3 MT246484_3 MT246482_2 MT246481_3 MT246480_1 MT246479_3 MT246478_1 MT246477_1 MT246476_2 MT246475_3 MT246474_1 MT246473_3 MT246472_1 MT246471_3 MT246469_1 MT246468_2 MT246467_2 MT246466_2 MT246464_2 MT246462_2 MT246461_3 MT246460_2 MT246459_2 MT246458_2 MT246457_3 MT246455_2 MT246454_2 MT246453_3 MT246452_1 MT246450_1 MT246449_3 are the most similar ones by 100.000000 percent identity\n",
      "MT233522_2 MT246470_2 MT246451_2 are the most different ones by 99.524000 percent identity\n"
     ]
    }
   ],
   "source": [
    "# Q2. Which pairs are most similar ones? Which pairs are most different ones?\n",
    "\n",
    "# in cDNA\n",
    "print('In terms of cDNA')\n",
    "try:\n",
    "    f = open('2019nCoV_genomes.2020_03_27_blastn.out', 'r')\n",
    "\n",
    "    subject_list = []\n",
    "    identity_list = []\n",
    "    for line in f:\n",
    "        subject_list.append(line.split()[1])\n",
    "        identity_list.append(float(line.split()[2]))\n",
    "\n",
    "    for i in range(len(identity_list)):\n",
    "        if identity_list[i] == max(identity_list):\n",
    "            print(subject_list[i], end = ' ')\n",
    "    print('are the most similar ones by %f percent identity' % max(identity_list))\n",
    "\n",
    "    for i in range(len(identity_list)):\n",
    "        if identity_list[i] == min(identity_list):\n",
    "            print(subject_list[i], end = ' ')\n",
    "    print('are the most different ones by %f percent identity' % min(identity_list))\n",
    "\n",
    "    f.close()\n",
    "except:\n",
    "    print('blast fail')\n",
    "\n",
    "print('\\n')\n",
    "\n",
    "# in protein\n",
    "print('In terms of protein')\n",
    "try:\n",
    "    f = open('2019nCoV_genomes.2020_03_27_trans6_blastp.out', 'r')\n",
    "\n",
    "    subject_list = []\n",
    "    identity_list = []\n",
    "    for line in f:\n",
    "        subject_list.append(line.split()[1])\n",
    "        identity_list.append(float(line.split()[2]))\n",
    "\n",
    "    for i in range(len(identity_list)):\n",
    "        if identity_list[i] == max(identity_list):\n",
    "            print(subject_list[i], end = ' ')\n",
    "    print('are the most similar ones by %f percent identity' % max(identity_list))\n",
    "\n",
    "    for i in range(len(identity_list)):\n",
    "        if identity_list[i] == min(identity_list):\n",
    "            print(subject_list[i], end = ' ')\n",
    "    print('are the most different ones by %f percent identity' % min(identity_list))\n",
    "\n",
    "\n",
    "    f.close()\n",
    "except:\n",
    "    print('blast fail')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Based on this info, extract N proteins cDNA/proteins from other SARS & MERS genomes.\n",
    "\n",
    "# SARS\n",
    "# in cDNA\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ makeblastdb -dbtype nucl -in ../data/SARS_genomes.2020_02_03.fa -out SARS_genomes.2020_02_03\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ blastn -db SARS_genomes.2020_02_03 -query N_genomic.fa -outfmt 6 -out SARS_genomes.2020_02_03_blastn.out\n",
    "\n",
    "f = open('SARS_genomes.2020_02_03_blastn.out', 'r')\n",
    "\n",
    "genome_list = read_fasta('../data/SARS_genomes.2020_02_03.fa')\n",
    "for line in f:\n",
    "    for genome in list(genome_list.keys()):\n",
    "        if line.split()[1] in genome:\n",
    "            print(end='')\n",
    "#            print(genome_list[genome][int(line.split()[8])-1:int(line.split()[9])], 'from', genome.split()[0])\n",
    "    \n",
    "f.close()\n",
    "\n",
    "\n",
    "# in protein\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ transeq --sequence ../data/SARS_genomes.2020_02_03.fa --outseq ../data/SARS_genomes.2020_02_03_trans6.fa -frame 6\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ makeblastdb -dbtype prot -in ../data/SARS_genomes.2020_02_03_trans6.fa -out ../data/SARS_genomes.2020_02_03_trans6\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ blastp -db ../data/SARS_genomes.2020_02_03_trans6 -query N_trans.fa -outfmt 6 -out SARS_genomes.2020_02_03_trans6_blastp.out\n",
    "\n",
    "f = open('SARS_genomes.2020_02_03_trans6_blastp.out', 'r')\n",
    "\n",
    "genome_list = read_fasta('../data/SARS_genomes.2020_02_03.fa')\n",
    "for line in f:\n",
    "    for genome in list(genome_list.keys()):\n",
    "        if line.split()[1] in genome:\n",
    "            print(end='')\n",
    "#            print(genome_list[genome][int(line.split()[8])-1:int(line.split()[9])], 'from', genome.split()[0])\n",
    "    \n",
    "f.close()\n",
    "\n",
    "\n",
    "# MERS\n",
    "# in cDNA\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ makeblastdb -dbtype nucl -in ../data/MERS_genomes.2020_02_03.fa -out MERS_genomes.2020_02_03\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ blastn -db MERS_genomes.2020_02_03 -query N_genomic.fa -outfmt 6 -out MERS_genomes.2020_02_03_blastn.out\n",
    "\n",
    "f = open('MERS_genomes.2020_02_03_blastn.out', 'r')\n",
    "\n",
    "genome_list = read_fasta('../data/MERS_genomes.2020_02_03.fa')\n",
    "for line in f:\n",
    "    for genome in list(genome_list.keys()):\n",
    "        if line.split()[1] in genome:\n",
    "            print(end='')\n",
    "#            print(genome_list[genome][int(line.split()[8])-1:int(line.split()[9])], 'from', genome.split()[0])\n",
    "    \n",
    "f.close()\n",
    "\n",
    "\n",
    "# in protein\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ transeq --sequence ../data/MERS_genomes.2020_02_03.fa --outseq ../data/MERS_genomes.2020_02_03_trans6.fa -frame 6\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ makeblastdb -dbtype prot -in ../data/MERS_genomes.2020_02_03_trans6.fa -out ../data/MERS_genomes.2020_02_03_trans6\n",
    "# (terminal) ~/git/BME203-2020/03_Annotation$ blastp -db ../data/MERS_genomes.2020_02_03_trans6 -query N_trans.fa -outfmt 6 -out MERS_genomes.2020_02_03_trans6_blastp.out\n",
    "\n",
    "f = open('MERS_genomes.2020_02_03_trans6_blastp.out', 'r')\n",
    "\n",
    "genome_list = read_fasta('../data/MERS_genomes.2020_02_03.fa')\n",
    "for line in f:\n",
    "    for genome in list(genome_list.keys()):\n",
    "        if line.split()[1] in genome:\n",
    "            print(end='')\n",
    "#            print(genome_list[genome][int(line.split()[8])-1:int(line.split()[9])], 'from', genome.split()[0])\n",
    "    \n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For SARS\n",
      "in terms of cDNA,\n",
      "ENA|DQ084200|DQ084200.1 ENA|DQ084199|DQ084199.1 ENA|DQ022305|DQ022305.2 are the most similar ones by 88.862000 percent identity\n",
      "ENA|AY686863|AY686863.1 are the most different ones by 88.217000 percent identity\n",
      "\n",
      "\n",
      "in terms of protein,\n",
      "KU973692.1_3 DQ898174.1_1 FJ882949.1_1 FJ882934.1_3 FJ882937.1_2 FJ882935.1_1 FJ882933.1_2 FJ882947.1_3 FJ882946.1_1 FJ882927.1_2 FJ882932.1_3 FJ882939.1_3 FJ882936.1_3 FJ882938.1_2 FJ882941.1_1 FJ882929.1_1 FJ882944.1_3 FJ882940.1_3 FJ882954.1_2 FJ882960.1_3 FJ882955.1_3 FJ882931.1_3 FJ882928.1_2 FJ882926.1_2 FJ882930.1_2 AY362698.1_1 AY394850.2_2 are the most similar ones by 90.544000 percent identity\n",
      "DQ084200.1_3 DQ022305.2_2 are the most different ones by 89.573000 percent identity\n",
      "\n",
      "\n",
      "\n",
      "For MERS\n",
      "in terms of cDNA,\n",
      "blast fail\n",
      "\n",
      "\n",
      "in terms of protein,\n",
      "KX574227.1_3 are the most similar ones by 50.142000 percent identity\n",
      "KX108941.1_2 are the most different ones by 43.333000 percent identity\n"
     ]
    }
   ],
   "source": [
    "# Q3. Which protein is most similar to SARS-CoV N protein? Which protein is most different to MERS N protein?\n",
    "# SARS\n",
    "print('For SARS')\n",
    "# in cDNA\n",
    "print('in terms of cDNA,')\n",
    "try:\n",
    "    f = open('SARS_genomes.2020_02_03_blastn.out', 'r')\n",
    "\n",
    "    subject_list = []\n",
    "    identity_list = []\n",
    "    for line in f:\n",
    "        subject_list.append(line.split()[1])\n",
    "        identity_list.append(float(line.split()[2]))\n",
    "\n",
    "    for i in range(len(identity_list)):\n",
    "        if identity_list[i] == max(identity_list):\n",
    "            print(subject_list[i], end = ' ')\n",
    "    print('are the most similar ones by %f percent identity' % max(identity_list))\n",
    "\n",
    "    for i in range(len(identity_list)):\n",
    "        if identity_list[i] == min(identity_list):\n",
    "            print(subject_list[i], end = ' ')\n",
    "    print('are the most different ones by %f percent identity' % min(identity_list))\n",
    "\n",
    "    f.close()\n",
    "except:\n",
    "    print('blast fail')\n",
    "\n",
    "print('\\n')\n",
    "\n",
    "# in protein\n",
    "print('in terms of protein,')\n",
    "try:\n",
    "    f = open('SARS_genomes.2020_02_03_trans6_blastp.out', 'r')\n",
    "\n",
    "    subject_list = []\n",
    "    identity_list = []\n",
    "    for line in f:\n",
    "        subject_list.append(line.split()[1])\n",
    "        identity_list.append(float(line.split()[2]))\n",
    "\n",
    "    for i in range(len(identity_list)):\n",
    "        if identity_list[i] == max(identity_list):\n",
    "            print(subject_list[i], end = ' ')\n",
    "    print('are the most similar ones by %f percent identity' % max(identity_list))\n",
    "\n",
    "    for i in range(len(identity_list)):\n",
    "        if identity_list[i] == min(identity_list):\n",
    "            print(subject_list[i], end = ' ')\n",
    "    print('are the most different ones by %f percent identity' % min(identity_list))\n",
    "\n",
    "    f.close()\n",
    "except:\n",
    "    print('blast fail')\n",
    "\n",
    "print('\\n\\n')\n",
    "\n",
    "\n",
    "# MERS\n",
    "print('For MERS')\n",
    "# in cDNA\n",
    "print('in terms of cDNA,')\n",
    "try:\n",
    "    f = open('MERS_genomes.2020_02_03_blastn.out', 'r')\n",
    "\n",
    "    subject_list = []\n",
    "    identity_list = []\n",
    "    for line in f:\n",
    "        subject_list.append(line.split()[1])\n",
    "        identity_list.append(float(line.split()[2]))\n",
    "        \n",
    "    for i in range(len(identity_list)):\n",
    "        if identity_list[i] == max(identity_list):\n",
    "            print(subject_list[i], end = ' ')\n",
    "    print('are the most similar ones by %f percent identity' % max(identity_list))\n",
    "\n",
    "    for i in range(len(identity_list)):\n",
    "        if identity_list[i] == min(identity_list):\n",
    "            print(subject_list[i], end = ' ')\n",
    "    print('are the most different ones by %f percent identity' % min(identity_list))\n",
    "    \n",
    "    f.close()\n",
    "except:\n",
    "    print('blast fail')\n",
    "\n",
    "print('\\n')\n",
    "\n",
    "# in protein\n",
    "print('in terms of protein,')\n",
    "try:\n",
    "    f = open('MERS_genomes.2020_02_03_trans6_blastp.out', 'r')\n",
    "\n",
    "    subject_list = []\n",
    "    identity_list = []\n",
    "    for line in f:\n",
    "        subject_list.append(line.split()[1])\n",
    "        identity_list.append(float(line.split()[2]))\n",
    "\n",
    "    for i in range(len(identity_list)):\n",
    "        if identity_list[i] == max(identity_list):\n",
    "            print(subject_list[i], end = ' ')\n",
    "    print('are the most similar ones by %f percent identity' % max(identity_list))\n",
    "\n",
    "    for i in range(len(identity_list)):\n",
    "        if identity_list[i] == min(identity_list):\n",
    "            print(subject_list[i], end = ' ')\n",
    "    print('are the most different ones by %f percent identity' % min(identity_list))\n",
    "\n",
    "    f.close()\n",
    "except:\n",
    "    print('blast fail')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
