{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homework 03. Jeehee Kim (Student ID 20205098)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read a FASTA file\n",
    "def read_fasta(tmp_filename):\n",
    "    rv = dict()\n",
    "    f = open(tmp_filename, 'r')\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            tmp_h = line.strip().lstrip('>')\n",
    "            rv[tmp_h] = ''\n",
    "        else:\n",
    "            rv[tmp_h] += line.strip().replace(' ', '')\n",
    "    f.close()\n",
    "    return rv\n",
    "\n",
    "# Make the fuction for transfer dictionary format to fasta format\n",
    "def w_fasta(fa_filename, g_dict):\n",
    "    fa_F = open(fa_filename, 'w')\n",
    "    for g_hd,g_sq in g_dict.items():\n",
    "        fa_F.write(\">\"+g_hd+\"\\n\"+g_sq+\"\\n\")        \n",
    "    fa_F.close()\n",
    "    \n",
    "###########################################################################   \n",
    "## 1. Read GCF_009858895.2 genomic info.\n",
    "\n",
    "GCF_f = '../data/GCF_009858895.2_ASM985889v3_genomic.fna'\n",
    "\n",
    "Ref_seq = read_fasta(GCF_f)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(28273, 29532)\n",
      " \n",
      "N protein:  ATGTCTGATAATGGACCCCAAAATCAGCGA\n"
     ]
    }
   ],
   "source": [
    "## 2. Make N-protein cDNA sequence.\n",
    "### N-protein index in GCF_009858895.2\n",
    "f_ref_idx = '../data/GCF_009858895.2_ASM985889v3_genomic.gff'\n",
    "\n",
    "def read_gff(tmp_filename, gene_name):\n",
    "    f = open(tmp_filename, 'r')\n",
    "    for line in f.readlines():\n",
    "        if 'gene' in line and \"gene={}\".format(gene_name) in line:\n",
    "            return int(line.split()[3])-1, int(line.split()[4])-1\n",
    "        \n",
    "\n",
    "print(read_gff(f_ref_idx, 'N'))\n",
    "### Get N-protein cDNA in fasta form\n",
    "\n",
    "Idx_N = read_gff(f_ref_idx, 'N')\n",
    "print(' ')\n",
    "\n",
    "for seq in Ref_seq.values():\n",
    "    ref_N_seq = seq[Idx_N[0] : Idx_N[1]+1]\n",
    "ref_N_cDNA ={'N_protein': ref_N_seq}\n",
    "#print(ref_N_cDNA)\n",
    "print('N protein: ', ref_N_cDNA['N_protein'][:30])\n",
    "\n",
    "w_fasta(\"ref_N_cDNA.fa\", ref_N_cDNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "They are same!\n"
     ]
    }
   ],
   "source": [
    "## 3. Translate it to protein sequence.  \n",
    "#transeq --sequence ../02_PCR/ref_N_cDNA.fa --outseq ref_N_protein.fa\n",
    "\n",
    "ref_N_protein = []\n",
    "\n",
    "ref_P = read_fasta('ref_N_protein.fa')\n",
    "for H_P,S_P in ref_P.items():\n",
    "    ref_N_protein.append(S_P)\n",
    "\n",
    "#print(ref_N_protein[0])\n",
    "\n",
    "###########################################################################\n",
    "## Q1. Compare it to GCF_009858895.2_ASM985889v3_protein.fna. Are they same?\n",
    "Sample_protein = '../data/GCF_009858895.2_ASM985889v3_protein.faa'\n",
    "sampleP_file = read_fasta(Sample_protein)\n",
    "#print(sampleP_file)\n",
    "\n",
    "Sample_N_protein = []\n",
    "for h_P, s_P in sampleP_file.items():\n",
    "    if 'nucleocapsid phosphoprotein' in h_P:\n",
    "        #print(s_P)\n",
    "        Sample_N_protein.append(s_P)\n",
    "        \n",
    "\n",
    "if ref_N_protein[0][:-1] == Sample_N_protein[0]:     # [:-1] ; for stop codon\n",
    "    print('They are same!')\n",
    "else:\n",
    "    print('No!!!!!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Perpectly matched! MT246454_1\n",
      "Perpectly matched! MT123291_1\n",
      "Perpectly matched! MT123293_1\n",
      "Perpectly matched! MT246449_1\n",
      "Perpectly matched! MT246450_1\n",
      "Perpectly matched! MT246452_1\n",
      "Perpectly matched! MT246453_1\n",
      "Perpectly matched! MT246455_1\n",
      "Perpectly matched! MT246457_1\n",
      "Perpectly matched! MT246458_1\n",
      "Perpectly matched! MT246459_1\n",
      "Perpectly matched! MT246460_1\n",
      "Perpectly matched! MT246461_1\n",
      "Perpectly matched! MT246462_1\n",
      "Perpectly matched! MT246464_1\n",
      "Perpectly matched! MT246466_1\n",
      "Perpectly matched! MT246467_1\n",
      "Perpectly matched! MT246468_1\n",
      "Perpectly matched! MT246469_1\n",
      "Perpectly matched! MT246471_1\n",
      "Perpectly matched! MT246472_1\n",
      "Perpectly matched! MT246473_1\n",
      "Perpectly matched! MT246474_1\n",
      "Perpectly matched! MT246475_1\n",
      "Perpectly matched! MT246476_1\n",
      "Perpectly matched! MT246477_1\n",
      "Perpectly matched! MT246478_1\n",
      "Perpectly matched! MT246479_1\n",
      "Perpectly matched! MT246480_1\n",
      "Perpectly matched! MT246481_1\n",
      "Perpectly matched! MT246482_1\n",
      "Perpectly matched! MT246484_1\n",
      "Perpectly matched! MT246486_1\n",
      "Perpectly matched! MT246487_1\n",
      "Perpectly matched! MT246488_1\n",
      "Perpectly matched! MT246489_1\n",
      "Perpectly matched! MT246490_1\n",
      "Perpectly matched! MT240479_1\n",
      "Perpectly matched! MT192772_1\n",
      "Perpectly matched! MT192765_1\n",
      "Perpectly matched! MT192759_1\n",
      "Perpectly matched! MT192773_1\n",
      "Perpectly matched! MT163719_1\n",
      "Perpectly matched! MT163716_1\n",
      "Perpectly matched! MT163717_1\n",
      "Perpectly matched! MT163718_1\n",
      "Perpectly matched! MT121215_1\n",
      "Perpectly matched! MT066156_1\n",
      "Perpectly matched! MT012098_1\n",
      "Perpectly matched! MT050493_1\n",
      "Perpectly matched! MT152824_1\n",
      "Perpectly matched! MT135041_1\n",
      "Perpectly matched! MT135042_1\n",
      "Perpectly matched! MT135044_1\n",
      "Perpectly matched! MT126808_1\n",
      "Perpectly matched! LC528233_1\n",
      "Perpectly matched! LC528232_1\n",
      "Perpectly matched! MT123292_1\n",
      "Perpectly matched! MT093631_1\n",
      "Perpectly matched! MT093571_1\n",
      "Perpectly matched! MT072688_1\n",
      "Perpectly matched! MT066175_1\n",
      "Perpectly matched! MT066176_1\n",
      "Perpectly matched! MT049951_1\n",
      "MT246456_1 99.762 % matched. 1 mismatches\n",
      "MT246485_1 99.762 % matched. 1 mismatches\n",
      "MT123290_1 99.762 % matched. 1 mismatches\n",
      "MT233519_1 99.762 % matched. 1 mismatches\n",
      "MT233523_1 99.762 % matched. 1 mismatches\n",
      "MT135043_1 99.762 % matched. 1 mismatches\n",
      "MT246451_1 99.524 % matched. 2 mismatches\n",
      "MT246470_1 99.524 % matched. 2 mismatches\n",
      "MT233522_1 99.524 % matched. 2 mismatches\n",
      " \n",
      "SARS-CoV-2 genomes compare with N protein's cDNA\n",
      "Most matched: 100.0 ['MT049951', 'MT066176', 'MT066175', 'MT072688', 'MT093571', 'MT093631', 'MT123292', 'LC528232', 'LC528233', 'MT126808', 'MT135044', 'MT135042', 'MT135041', 'MT152824', 'MT050493', 'MT012098', 'MT066156', 'MT121215', 'MT163718', 'MT163717', 'MT163716', 'MT163719', 'MT192773', 'MT192759', 'MT192765', 'MT192772', 'MT240479', 'MT246490', 'MT246489', 'MT246488', 'MT246487', 'MT246486', 'MT246484', 'MT246482', 'MT246481', 'MT246480', 'MT246479', 'MT246478', 'MT246477', 'MT246476', 'MT246475', 'MT246474', 'MT246473', 'MT246472', 'MT246471', 'MT246469', 'MT246468', 'MT246467', 'MT246466', 'MT246464', 'MT246462', 'MT246461', 'MT246460', 'MT246459', 'MT246458', 'MT246457', 'MT246455', 'MT246453', 'MT246452', 'MT246450', 'MT246449']\n",
      "Most different: 99.762 ['MT233522', 'MT246470', 'MT246451']\n",
      " \n",
      "SARS-CoV-2 genomes compare with N protein's amino acid seq\n",
      "Most matched: 100.0 ['MT246454_1', 'MT123291_1', 'MT123293_1', 'MT246449_1', 'MT246450_1', 'MT246452_1', 'MT246453_1', 'MT246455_1', 'MT246457_1', 'MT246458_1', 'MT246459_1', 'MT246460_1', 'MT246461_1', 'MT246462_1', 'MT246464_1', 'MT246466_1', 'MT246467_1', 'MT246468_1', 'MT246469_1', 'MT246471_1', 'MT246472_1', 'MT246473_1', 'MT246474_1', 'MT246475_1', 'MT246476_1', 'MT246477_1', 'MT246478_1', 'MT246479_1', 'MT246480_1', 'MT246481_1', 'MT246482_1', 'MT246484_1', 'MT246486_1', 'MT246487_1', 'MT246488_1', 'MT246489_1', 'MT246490_1', 'MT240479_1', 'MT192772_1', 'MT192765_1', 'MT192759_1', 'MT192773_1', 'MT163719_1', 'MT163716_1', 'MT163717_1', 'MT163718_1', 'MT121215_1', 'MT066156_1', 'MT012098_1', 'MT050493_1', 'MT152824_1', 'MT135041_1', 'MT135042_1', 'MT135044_1', 'MT126808_1', 'LC528233_1', 'LC528232_1', 'MT123292_1', 'MT093631_1', 'MT093571_1', 'MT072688_1', 'MT066175_1', 'MT066176_1', 'MT049951_1']\n",
      "Most different: 99.524 ['MT246451_1', 'MT246470_1', 'MT233522_1']\n"
     ]
    }
   ],
   "source": [
    "## 4. Based on this info, extract N proteins cDNA/proteins from other SARS-CoV-2 genomes (either 02_03 or 03_27).\n",
    "## Compare reference N protein cDNA with SARS-CoV-genome using by blastn\n",
    "#makeblastdb -dbtype nucl -in ../data/2019nCoV_genomes.2020_03_27.fa -out 2019nCoV_genomes.2020_03_27.blastnb.nbl\n",
    "#blastn -db 2019nCoV_genomes.2020_03_27.blastnb.nbl -query ref_N_cDNA.fa -outfmt 7 -out cDNA_compare.nblast.nbl\n",
    "\n",
    "### Extract N protein CDS form 2019nCoV_genomes\n",
    "g200327_f = '../data/2019nCoV_genomes.2020_03_27.fa'\n",
    "rd_g200327_dict = read_fasta(g200327_f)\n",
    "\n",
    "def Extr_P_seq(nblasted_file,query,seq_dict):\n",
    "    protein_idx = dict()\n",
    "    f = open(nblasted_file, 'r')\n",
    "    for line in f.readlines():\n",
    "        if line.startswith(query):\n",
    "            s_h = str(line.split()[1])\n",
    "            idx = (int(line.split()[8])-1), (int(line.split()[9])-1)            \n",
    "            protein_idx[s_h]=seq_dict[s_h][idx[0]:idx[1]+1] \n",
    "    return protein_idx\n",
    "   \n",
    "    \n",
    "k = './cDNA_compare.nblast.nbl'\n",
    "s_list = '../data/2019nCoV_genomes.2020_03_27.fa'\n",
    "\n",
    "S = read_fasta(s_list)\n",
    "s_dict = dict()\n",
    "for H0327, S0327 in S.items():\n",
    "    s_dict.setdefault(H0327[:8],S0327)\n",
    "\n",
    "Nseq_dict = Extr_P_seq( k , 'N_protein' , s_dict)\n",
    "#print(s_dict)\n",
    "#print(Nseq_dict)\n",
    "\n",
    "\n",
    "t = w_fasta('./200327_N_seq.fa', Nseq_dict)\n",
    "\n",
    "## Translate it to protein sequence.\n",
    "#transeq --sequence 200327_N_seq.fa --outseq 200327_N_AA.fa\n",
    "## Compare ref_N_protein amino acid with SARS-CoV-AA using by blastp\n",
    "#makeblastdb -dbtype prot -in 200327_N_AA.fa -out 200327_N_protein.pblast\n",
    "#blastp -db 200327_N_protein.pblast -query ref_N_protein.fa -outfmt 7 -out compare_200327SARS_CoV_2.blastp.tbl\n",
    "\n",
    "def read_blast(blasted_file,query):\n",
    "    f = open(blasted_file,'r')\n",
    "    for line in f.readlines():\n",
    "        if line.startswith(query):\n",
    "            if float(line.split()[2]) == 100.000 :\n",
    "                s_name = str(line.split()[1])\n",
    "                print(\"Perpectly matched!\", s_name)\n",
    "            else:\n",
    "                s_name = str(line.split()[1])\n",
    "                Pcnt = str(line.split()[2]) \n",
    "                mismatch = str(line.split()[4])\n",
    "                print(s_name, Pcnt, \"% matched.\",mismatch,\"mismatches\")\n",
    "\n",
    "read_blast(\"./compare_200327SARS_CoV_2.blastp.tbl\", \"N_protein_1\")\n",
    "print(' ')\n",
    "\n",
    "## Similarity checking\n",
    "def blast_similarity(blasted_file,query):\n",
    "    f = open(blasted_file,'r')\n",
    "    percentage_list = []\n",
    "    for line in f.readlines():\n",
    "        if line.startswith(query):\n",
    "            percentage_list.append(float(line.split()[2]))\n",
    "    max_pcntg = max(percentage_list)\n",
    "    min_pcntg = min(percentage_list)\n",
    "    f.close()\n",
    "    \n",
    "    f = open(blasted_file,'r')\n",
    "    max_pcntg_name = []\n",
    "    min_pcntg_name = []\n",
    "    for line in f.readlines():\n",
    "        if line.startswith(query):\n",
    "            if float(line.split()[2]) == max_pcntg:\n",
    "                max_pcntg_name.append(line.split()[1])\n",
    "            elif float(line.split()[2]) == min_pcntg:\n",
    "                min_pcntg_name.append(line.split()[1])\n",
    "    f.close()        \n",
    "    print( \"Most matched:\",max_pcntg, max_pcntg_name )\n",
    "    print( \"Most different:\",min_pcntg, min_pcntg_name )\n",
    "    \n",
    "# Q2. Which pairs are most similar ones? Which pairs are most different ones?\n",
    "print(\"SARS-CoV-2 genomes compare with N protein's cDNA\")\n",
    "blast_similarity('cDNA_compare.nblast.nbl','N_protein')\n",
    "print(' ')\n",
    "print(\"SARS-CoV-2 genomes compare with N protein's amino acid seq\")\n",
    "blast_similarity('./compare_200327SARS_CoV_2.blastp.tbl','N_protein_1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      "SARS proteins compare with N protein's amino acid seq\n",
      "Most matched: 90.544 ['KU973692.1_3', 'DQ898174.1_1', 'FJ882949.1_1', 'FJ882934.1_3', 'FJ882937.1_2', 'FJ882935.1_1', 'FJ882933.1_2', 'FJ882947.1_3', 'FJ882946.1_1', 'FJ882927.1_2', 'FJ882932.1_3', 'FJ882939.1_3', 'FJ882936.1_3', 'FJ882938.1_2', 'FJ882941.1_1', 'FJ882929.1_1', 'FJ882944.1_3', 'FJ882940.1_3', 'FJ882954.1_2', 'FJ882960.1_3', 'FJ882955.1_3', 'FJ882931.1_3', 'FJ882928.1_2', 'FJ882926.1_2', 'FJ882930.1_2', 'AY362698.1_1', 'AY394850.2_2']\n",
      "Most different: 89.573 ['DQ084200.1_3', 'DQ022305.2_2']\n",
      " \n",
      "MERS proteins compare with N protein's amino acid seq\n",
      "Most matched: 50.142 ['KX574227.1_3']\n",
      "Most different: 43.333 ['KX108941.1_2']\n"
     ]
    }
   ],
   "source": [
    "# 5. Based on this info, extract N proteins cDNA/proteins from other SARS & MERS genomes.\n",
    "#makeblastdb -dbtype nucl -in ../data/SARS_genomes.2020_02_03.fa -out SARS_genomes.2020_02_03.blastn.nbl\n",
    "#blastn -db SARS_genomes.2020_02_03.blastn.nbl -query ref_N_cDNA.fa -outfmt 7 -out SARS_genome_compare.blastn.nbl\n",
    "\n",
    "SARS_g = '../data/SARS_genomes.2020_02_03.fa'\n",
    "#transeq --sequence ../data/SARS_genomes.2020_02_03.fa --outseq SARS_genome.2020_02_03_AA.fa -frame 6\n",
    "#makeblastdb -dbtype prot -in ../data/SARS_genome.2020_02_03_AA.fa -out SARS_genome_200203.blastp\n",
    "#blastp -db SARS_genome_200203.blastp -query ref_N_protein.fa -outfmt 7 -out compare_200203SARS.blastp.tbl\n",
    "\n",
    "MERS_g = '../data/MERS_genomes.2020_02_03.fa'\n",
    "#transeq --sequence ../data/MERS_genomes.2020_02_03.fa --outseq MERS_genome.2020_02_03_AA.fa -frame 6\n",
    "#makeblastdb -dbtype prot -in MERS_genome.2020_02_03_AA.fa -out MERS_genome_200203.blastp\n",
    "#blastp -db MERS_genome_200203.blastp -query ref_N_protein.fa -outfmt 7 -out compare_200203MERS.blastp.tbl\n",
    "\n",
    "\n",
    "# Q3. Which protein is most similar to SARS-CoV N protein? Which protein is most different to MERS N protein?\n",
    "print(' ')\n",
    "print(\"SARS proteins compare with N protein's amino acid seq\")\n",
    "blast_similarity('compare_200203SARS.blastp.tbl','N_protein_1')\n",
    "\n",
    "print(' ')\n",
    "print(\"MERS proteins compare with N protein's amino acid seq\")\n",
    "blast_similarity('compare_200203MERS.blastp.tbl','N_protein_1')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
