{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homework 04. Chanwoo Kim (student ID:20185288)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3\n",
    "\n",
    "seq_info = dict()\n",
    "f_csv = open('2019nCoV_genomes.2020_04_14.csv', 'r')\n",
    "f_csv.readline()\n",
    "for line in f_csv:\n",
    "    tokens = line.replace(' ', '_').strip().split(',')\n",
    "    seq_id = tokens[0]\n",
    "\n",
    "    collect_date = tokens[-1]\n",
    "    if len(collect_date) == 7:\n",
    "        collect_date += '-00'\n",
    "    geo_location = tokens[4]\n",
    "    geo_location = geo_location.replace('North_Carolina', 'NC')\n",
    "    geo_location = geo_location.replace('Illinois', 'IL')\n",
    "    geo_location = geo_location.replace('San_Francisco', 'CA-SF')\n",
    "    geo_location = geo_location.replace('\"', '')\n",
    "    isolation_source = tokens[6]\n",
    "    #print(geo_location)\n",
    "    if geo_location not in seq_info:\n",
    "        seq_info[geo_location] = dict()\n",
    "\n",
    "    seq_info[geo_location][seq_id] = collect_date\n",
    "f_csv.close()\n",
    "#Accession,Release_Date,Species,Length,Geo_Location,Host,Isolation_Source,Collection_Date\n",
    "#NC_045512,2020-01-13T00:00:00Z,Severe acute respiratory syndrome-related coronavirus,29903,China,Homo sapiens,,2019-12\n",
    "\n",
    "#==> 2019nCoV_genomes.2020_04_14.fa <==\n",
    "#>NC_045512 |Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1| complete genome\n",
    "#ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCT\n",
    "\n",
    "seq_list = dict()\n",
    "f_fa = open('../data/2019nCoV_genomes.2020_04_14.fa', 'r')\n",
    "for line in f_fa:\n",
    "    if line.startswith('>'):\n",
    "        tmp_h = line.strip().lstrip('>').split()[0]\n",
    "        seq_list[tmp_h] = []\n",
    "    else:\n",
    "        seq_list[tmp_h].append(line.strip())\n",
    "f_fa.close()\n",
    "\n",
    "for tmp_loc in sorted(seq_info.keys()):\n",
    "    if tmp_loc == '':\n",
    "        continue\n",
    "    if len(seq_info[tmp_loc]) < 3:\n",
    "        continue\n",
    "\n",
    "    for tmp_id, tmp_date in seq_info[tmp_loc].items():\n",
    "        tmp_seq = ''.join(seq_list[tmp_id])\n",
    "        print(\">%s|%s|%s\\n%s\" % (tmp_id, tmp_loc.replace(':', '_'), tmp_date, tmp_seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#report code\n",
    "\n",
    "# The modified function to read sequences from a FASTA file for ease of use.\n",
    "def read_fasta(tmp_filename):\n",
    "    rv = dict()\n",
    "    f = open(tmp_filename, 'r')\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            tmp_h = line\n",
    "            rv[tmp_h] = ''\n",
    "        else:\n",
    "            rv[tmp_h] += line.strip().replace(' ', '')\n",
    "    f.close()\n",
    "    return rv\n",
    "\n",
    "#To make a fasta file that has only genomes from US.\n",
    "covid_genome= \"2019nCoV_genomes.2020_04_14.clean.fa\"\n",
    "genome1=read_fasta(covid_genome)\n",
    "USAgenome = open(\"2019CoV_USAgenomes.fa\",\"w\")\n",
    "for gid, gseq in genome1.items():\n",
    "    if \"USA\" in gid:\n",
    "        USAgenome.write(gid)\n",
    "        gseq=gseq+\"\\n\"\n",
    "        USAgenome.write(gseq)\n",
    "USAgenome.close()              \n",
    "\n",
    "f = open(\"2019nCoV_genomes.2020_04_14.csv\", \"r\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_genome= \"GCF_009858895.2_ASM985889v3_genomic.fasta\"\n",
    "GCF_009858895_genome=read_fasta(ref_genome)\n",
    "f = open(\"reference.fasta\", \"w\")\n",
    "for x,y in GCF_009858895_genome.items():\n",
    "    f.write(x)\n",
    "f.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This file filters the fasta input file to filter the genome sequences that have the geological location within USA\n",
    "# for multiple sequence alignment via nextstrain\n",
    "def read_fasta(tmp_filename):\n",
    "    rv = dict()\n",
    "    f = open(tmp_filename, 'r')\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            tmp_h = line\n",
    "            rv[tmp_h] = ''\n",
    "        else:\n",
    "            rv[tmp_h] += line.strip().replace(' ', '')\n",
    "    f.close()\n",
    "    return rv\n",
    "covid_genome= \"2019CoV_USAgenomes.fa\"\n",
    "genome=read_fasta(covid_genome)\n",
    "USAgenome = open(\"2019CoV_USAgenomes2.fa\",\"w\")\n",
    "for gid, gseq in genome.items():\n",
    "  \n",
    "    gid2=gid.split(\"|\")[0]+'\\n'\n",
    "    USAgenome.write(gid2)\n",
    "    gseq=gseq+\"\\n\"\n",
    "    USAgenome.write(gseq)\n",
    "USAgenome.close()   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Here, for the input file I used the sequence file of the CoV-2 genomes where all the genomes are aligned with mafft program via nextstrain.\n",
    "gf=open(\"GCF_009858895.2_ASM985889v3_genomic.gff\",\"r\")\n",
    "glines=gf.readlines()\n",
    "genename=set()\n",
    "for r in glines[7:-1]:\n",
    "    if r.split()[2]==\"gene\":\n",
    "        genename.add((r.split()[8].split(';')[2].split('=')[1]))\n",
    "\n",
    "alignedgenome=\"aligned.fasta\"\n",
    "genome=read_fasta(alignedgenome)\n",
    "#From gene information available in gff file, I extracted the sequence information available for each genome, and wrote the fasta file \n",
    "# for each aligned gene.\n",
    "for n in genename:\n",
    "    genealign=open(\"%saligned.fasta\" % (n),\"w\")\n",
    "    for r in glines[9:-2]:\n",
    "        if n==r.split()[8].split(';')[2].split('=')[1]:\n",
    "            for gid, gseq in genome.items():\n",
    "                geneid=gid.rstrip(\"\\n\")+n+'\\n'\n",
    "                genealign.write(geneid)\n",
    "                geneseq=gseq[int(r.split()[3])-1:int(r.split()[4])]+'\\n'\n",
    "                genealign.write(geneseq)\n",
    "    genealign.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "1\n",
      "1\n",
      "1\n",
      "1\n",
      "0\n",
      "1\n",
      "1\n",
      "1\n",
      "1\n",
      "1\n",
      "0\n",
      "0\n",
      "0\n",
      "0.5\n",
      "0.5\n",
      "0.5\n",
      "0.5\n",
      "0.5\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "2\n",
      "2\n",
      "2\n",
      "2\n",
      "2\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "54\n",
      "0.04205607476635514 is the diversity score for the protein/genome\n"
     ]
    }
   ],
   "source": [
    "#Sequence diversity plot code\n",
    "\n",
    "alignedgenome=\"ORF10aligned.fasta\" #file to align\n",
    "genome=read_fasta(alignedgenome)\n",
    "seqsize=len(genome['>MT258380\\n']) \n",
    "window_length=10    #size of the sliding window\n",
    "step=2                #the step size for each itegration of the sliding window shift.       \n",
    "base=str()\n",
    "total_score=0      #total diversity to calculate\n",
    "def my_range(start, end, Step):\n",
    "    while start <= end:\n",
    "        yield start\n",
    "        start += Step\n",
    "f=open(\"variabliltytableORF10window10.txt\",\"w\")                       #output file\n",
    "f.write(\"Nucleotide_position\"+'\\t'+'Variability_score'+'\\n')\n",
    "  #Multiple loops for the calculation of the diversity score at each sliding window\n",
    "for x in my_range(1, seqsize-window_length,step):             \n",
    "    score=0\n",
    "    f.write(\"%s\" % x+'\\t')\n",
    "    for y in  range(0,window_length):     #At each window shift, the base counts are reset to 0.\n",
    "        A=0\n",
    "        C=0\n",
    "        G=0\n",
    "        T=0\n",
    "        N=0\n",
    "        del base\n",
    "        z=x+y\n",
    "#process of how the consensus base is determined by counting each base occurrence\n",
    "        for gid,gseq in genome.items():         \n",
    "            if gseq[z]==\"A\":\n",
    "                A+=1\n",
    "            if gseq[z]==\"C\":\n",
    "                C+=1\n",
    "            if gseq[z]==\"G\":\n",
    "                G+=1\n",
    "            if gseq[z]==\"T\":\n",
    "                T+=1\n",
    "            if gseq[z]==\"N\":\n",
    "                N+=1\n",
    "            if gseq[z]==\"S\":\n",
    "                G+=1\n",
    "                C+=1\n",
    "            if gseq[z]==\"W\":\n",
    "                A+=1\n",
    "                T+=1\n",
    "            if gseq[z]==\"M\":\n",
    "                A+=1\n",
    "                C+=1\n",
    "            if gseq[z]==\"R\":\n",
    "                A+=1\n",
    "                G+=1            \n",
    "            if gseq[z]==\"Y\":\n",
    "                T+=1\n",
    "                C+=1\n",
    "            if gseq[z]==\"K\":\n",
    "                G+=1\n",
    "                T+=1\n",
    "            if gseq[z]==\"B\":\n",
    "                A-=1\n",
    "            if gseq[z]==\"D\":\n",
    "                C-=1\n",
    "            if gseq[z]==\"H\":\n",
    "                G-=1\n",
    "            if gseq[z]==\"V\":\n",
    "                T-=1\n",
    "        if max(A,C,G,T,N)==A:\n",
    "            base=\"A\"\n",
    "        if max(A,C,G,T,N)==C:\n",
    "            base=\"C\"\n",
    "        if max(A,C,G,T,N)==G:\n",
    "            base=\"G\"\n",
    "        if max(A,C,G,T,N)==T:\n",
    "            base=\"T\"\n",
    "        if max(A,C,G,T,N)==N:\n",
    "            base=\"N\"\n",
    "        \n",
    "#Code to get the score for each sliding window\n",
    "        for gid,gseq in genome.items():\n",
    "            if base!=gseq[z]:\n",
    "                score+=1\n",
    "            if base==\"N\" and base!=gseq[z]:\n",
    "                score-=1\n",
    "            if base==\"A\" and gseq[z]==(\"W\"or\"M\"or \"R\"):\n",
    "                score-=0.5\n",
    "            if base==\"C\" and gseq[z]==(\"S\" or \"M\" or \"Y\"):\n",
    "                score-=0.5\n",
    "            if base==\"G\" and gseq[z]==(\"S\" or\"K\"or\"R\"):\n",
    "                score-=0.5\n",
    "            if base==\"T\" and gseq[z]==(\"Y\" or\"K\"or \"W\"):\n",
    "                score-=0.5\n",
    "            if base==\"A\" and gseq[z]==(\"D\" or \"V\" or \"H\"):\n",
    "                score-=0.33\n",
    "            if base==\"C\" and gseq[z]==(\"B\"or \"V\"or\"H\"):\n",
    "                score-=0.33\n",
    "            if base==\"G\" and gseq[z]==(\"D\" or \"B\" or \"V\"):\n",
    "                score-=0.33\n",
    "            if base==\"T\" and gseq[z]==(\"D\" or \"H\" or \"V\"):\n",
    "                score-=0.33 \n",
    "                \n",
    "    f.write(str(score)+'\\n')    \n",
    "    scorelist.append(score)\n",
    "    total_score+=score\n",
    "    print(score)\n",
    "#This is the output of the code which produces the normalized diversity score for the analyzed protein.\n",
    "print(total_score/(seqsize-window_length)*step/window_length, \"is the normalized diversity score for the protein/genome\")\n",
    "\n",
    "f.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "analysis start\n",
      "5001 iterations\n",
      "5001 iterations\n",
      "5001 iterations\n",
      "5001 iterations\n",
      "10001 iterations\n",
      "10001 iterations\n",
      "10001 iterations\n",
      "10001 iterations\n",
      "20001 iterations\n",
      "20001 iterations\n",
      "20001 iterations\n",
      "20001 iterations\n",
      "{'N', 'T', 'G', 'A', 'C'}\n"
     ]
    }
   ],
   "source": [
    "#Just to find the set containing the consensus base.\n",
    "import math\n",
    "alignedgenome=\"aligned.fasta\"\n",
    "genome=read_fasta(alignedgenome)\n",
    "seqsize=len(genome['>MT258380\\n'])\n",
    "window_length=20\n",
    "step=5\n",
    "scorelist=list()\n",
    "baseset=set()\n",
    "\n",
    "def most_frequent(List): \n",
    "    counter = 0\n",
    "    num = List[0] \n",
    "      \n",
    "    for i in List: \n",
    "        curr_frequency = List.count(i) \n",
    "        if(curr_frequency> counter): \n",
    "            counter = curr_frequency \n",
    "            num = i \n",
    "  \n",
    "    return num \n",
    "\n",
    "def my_range(start, end, Step):\n",
    "    while start <= end:\n",
    "        yield start\n",
    "        start += Step\n",
    "        \n",
    "for x in my_range(1, seqsize-window_length,step):\n",
    "    score=0\n",
    "    for y in  range(0,window_length):\n",
    "        baselist=[]\n",
    "        z=x+y\n",
    "        if z+y==1:\n",
    "            print(\"analysis start\")        \n",
    "        if z+y==5001:\n",
    "            print(\"5001 iterations\")\n",
    "        if z+y==10001:\n",
    "            print(\"10001 iterations\")\n",
    "        if z+y==20001:\n",
    "            print(\"20001 iterations\")            \n",
    "        for gid,gseq in genome.items():\n",
    "            baselist.append(gseq[z])\n",
    "        baseset.add(most_frequent(baselist))\n",
    "#Output to show the possible consensus bases\n",
    "print(baseset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
