{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Report 05. Jeun Hyun (Student ID -20191328)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The function to read sequences from a FASTA file\n",
    "def read_fasta(tmp_filename):\n",
    "    rv = dict()\n",
    "    f = open(tmp_filename, 'r')\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            tmp_h = line.strip().lstrip('>')\n",
    "            rv[tmp_h] = ''\n",
    "        else:\n",
    "            rv[tmp_h] += line.strip().replace(' ', '')\n",
    "    f.close()\n",
    "    return rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make clean process\n",
    "genome = read_fasta('2019nCoV_genomes.2020_04_14.clean.fa')\n",
    "for i in genome.keys():\n",
    "    if len(i.split('|')[2]) == 4:\n",
    "        genome[i+'-00-00'] = genome[i]\n",
    "        break\n",
    "\n",
    "n_genome = {}\n",
    "for i in genome.keys():\n",
    "    n_genome[i.replace('|',';')] = genome[i]\n",
    "\n",
    "f = open(\"../data/2019nCoV_genomes.2020_04_14.clean_n.fa\", 'w')\n",
    "for name in n_genome.keys():\n",
    "    data = \"> %s\\n%s\\n\" %(name, n_genome[name])\n",
    "    f.write(data)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read GCF_009858895.2 genomic info.\n",
    "filename_genomes = '../data/GCF_009858895.2_ASM985889v3_genomic.fna'\n",
    "genome_list = read_fasta(filename_genomes)\n",
    "ref_seq = list(genome_list.values())[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make ref.orf1ab cDNA sequence.\n",
    "f = open('../data/GCF_009858895.2_ASM985889v3_genomic.gff', 'r')\n",
    "count = 1\n",
    "gene_dict = {}\n",
    "for line in f:\n",
    "    try:\n",
    "        if 'CDS' in line.split('\t')[2] and 'product=orf1ab polyprotein' in line.split('\t')[8]:\n",
    "            gene_dict['orf1ab_polyprotein_%s'%count] = ref_seq[int(line.split()[3])-1:int(line.split()[4])]\n",
    "            count +=1\n",
    "    except:\n",
    "        continue        \n",
    "f.close()\n",
    "\n",
    "# Make other CDN's cDNA ref.sequences.\n",
    "f = open('../data/GCF_009858895.2_ASM985889v3_genomic.gff', 'r')\n",
    "for line in f:\n",
    "    #print('line')\n",
    "    try:\n",
    "        #print('try')\n",
    "        if 'CDS' == line.split('\t')[2]:\n",
    "            info = line.split('\t')[8].split(';')\n",
    "            for h in info:\n",
    "                if 'product=' in h:\n",
    "                    name = h[8:]\n",
    "                    name_genomic = ref_seq[int(line.split()[3])-1 : int(line.split()[4])]\n",
    "                    #print('cDNA sequence of \\'%s gene\\' is'%name, name_genomic[:25], '...', name_genomic[-25:])\n",
    "                    if name != 'orf1ab polyprotein':\n",
    "                        gene_dict[name.replace(' ','_')] = name_genomic\n",
    "    except:\n",
    "        continue\n",
    "#print(list(gene_dict.keys()))\n",
    "f.close()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "f = open(\"../data/gene_dict.fa\", 'w')\n",
    "for name in gene_dict.keys():\n",
    "    #print(name)\n",
    "    data = \"> %s\\n%s\\n\" %(name, gene_dict[name])\n",
    "    f.write(data)\n",
    "f.close()\n",
    "\n",
    "\n",
    "# Do transeq and get reference protein sequences of each gene\n",
    "# Do blastp and get the location of protein sequences of subject fit into the reference sequence\n",
    "# (terminal) ~/git/BME203-2020$ transeq --sequence ./data/gene_dict.fa --outseq ./data/gene_dict_trans.fa\n",
    "# (terminal) ~/git/BME203-2020$ transeq --sequence ./data/2019nCoV_genomes.2020_04_14.clean_n.fa --outseq ./data/2019nCoV_genomes.2020_04_14.clean_n_transF.fa -frame F\n",
    "# (terminal) ~/git/BME203-2020$ makeblastdb -dbtype prot -in ./data/2019nCoV_genomes.2020_04_14.clean_n_transF.fa -out ./data/2019nCoV_genomes.2020_04_14.clean_n_transF\n",
    "# (terminal) ~/git/BME203-2020$ blastp -db ./data/2019nCoV_genomes.2020_04_14.clean_n_transF -query ./data/gene_dict_trans.fa -outfmt 6 -out ./data/2019nCoV_genomes.2020_04_14.clean_n_transF_blastp.out -max_target_seqs 100000 -max_hsps 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract the protein of each genes from each subjects.\n",
    "query_list = ['orf1ab_polyprotein_query','orf1a_polyprotein_query', 'surface_glycoprotein_query', 'ORF3a_protein_query', 'envelope_protein_query', 'membrane_glycoprotein_query' ,'ORF6_protein_query' ,'ORF7a_protein_query' ,'ORF7b_query', 'ORF8_protein_query', 'nucleocapsid_phosphoprotein_query', 'ORF10_protein_query']\n",
    "\n",
    "transF_list = read_fasta('../data/2019nCoV_genomes.2020_04_14.clean_n_transF.fa')\n",
    "temp = ()\n",
    "\n",
    "for query in query_list:\n",
    "    f = open('../data/2019nCoV_genomes.2020_04_14.clean_n_transF_blastp.out', 'r')\n",
    "    #print('query_list')\n",
    "    query_dict = {}\n",
    "    for line in f:\n",
    "        sep = line.split('\t')\n",
    "        #print(sep[0].split('_')[0], query)\n",
    "        if (sep[0],sep[1][:-2])!= temp:\n",
    "            if sep[0].split('_')[0] in query:\n",
    "                #print(sep[1][:-2],len(query_dict.keys()))\n",
    "                p_s = int(sep[8])-1\n",
    "                p_l = int(sep[9])\n",
    "                if sep[0][:-2] == 'orf1ab_polyprotein_1':\n",
    "                    query_dict[sep[1][:-2]] = transF_list[sep[1]][p_s:p_l]\n",
    "                    #print(p_s, p_l)\n",
    "                elif sep[0][:-2] == 'orf1ab_polyprotein_2':\n",
    "                    query_dict[sep[1][:-2]] = query_dict[sep[1][:-2]] + transF_list[sep[1]][p_s:p_l]\n",
    "                    #print(query_dict[sep[1]][:25])\n",
    "                else:\n",
    "                    query_dict[sep[1][:-2]] = transF_list[sep[1]][p_s:p_l]\n",
    "                    #print(query_dict[sep[1]][:25])\n",
    "                temp = (sep[0],sep[1][:-2])\n",
    "    \n",
    "\n",
    "    h = open(\"../data/%s_protein_dict.fa\"%query, 'w')\n",
    "    for name in query_dict.keys():\n",
    "        #print(name)\n",
    "        data = \"> %s\\n%s\\n\" %(name, query_dict[name])\n",
    "        h.write(data)\n",
    "    h.close() \n",
    "    f.close()\n",
    "\n",
    "# Do muscle and get the seuqnces matched amino acids with the same length\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/orf1ab_polyprotein_query_protein_dict.fa -out ./data/orf1ab_polyprotein_query_protein_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/orf1a_polyprotein_query_protein_dict.fa -out ./data/orf1a_polyprotein_query_protein_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/surface_glycoprotein_query_protein_dict.fa -out ./data/surface_glycoprotein_query_protein_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/ORF3a_protein_query_protein_dict.fa -out ./data/ORF3a_protein_query_protein_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/envelope_protein_query_protein_dict.fa -out ./data/envelope_protein_query_protein_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/membrane_glycoprotein_query_protein_dict.fa -out ./data/membrane_glycoprotein_query_protein_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/ORF6_protein_query_protein_dict.fa -out ./data/ORF6_protein_query_protein_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/ORF7a_protein_query_protein_dict.fa -out ./data/ORF7a_protein_query_protein_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/ORF7b_query_protein_dict.fa -out ./data/ORF7b_query_protein_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/ORF8_protein_query_protein_dict.fa -out ./data/ORF8_protein_query_protein_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/nucleocapsid_phosphoprotein_query_protein_dict.fa -out ./data/nucleocapsid_phosphoprotein_query_protein_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/ORF10_protein_query_protein_dict.fa -out ./data/ORF10_protein_query_protein_dict_muscle.fa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract the cDNA of each genes from each subjects.\n",
    "query_list = ['orf1ab_polyprotein_query','orf1a_polyprotein_query', 'surface_glycoprotein_query', 'ORF3a_protein_query', 'envelope_protein_query', 'membrane_glycoprotein_query' ,'ORF6_protein_query' ,'ORF7a_protein_query' ,'ORF7b_query', 'ORF8_protein_query', 'nucleocapsid_phosphoprotein_query', 'ORF10_protein_query']\n",
    "cDNA_list = read_fasta('../data/2019nCoV_genomes.2020_04_14.clean_n.fa')\n",
    "\n",
    "temp = ()\n",
    "for query in query_list:\n",
    "    f = open('../data/2019nCoV_genomes.2020_04_14.clean_n_transF_blastp.out', 'r')\n",
    "    query_dict = {}\n",
    "    for line in f:\n",
    "        sep = line.split('\t')\n",
    "        if (sep[0],sep[1][:-2])!= temp:\n",
    "            if sep[0].split('_')[0] in query:\n",
    "                #print(sep[1][:-2],len(query_dict.keys()))\n",
    "                p_s = int(sep[8])-1\n",
    "                p_l = int(sep[9])\n",
    "                if line.split()[1][-1:] == '1':\n",
    "                    if sep[0][:-2] == 'orf1ab_polyprotein_1':\n",
    "                        #print('_1')\n",
    "                        query_dict[sep[1][:-2]] = cDNA_list[' '+sep[1][:-2]][(int(p_s))*3 : int(p_l)*3]\n",
    "                        #print(int(p_s), (int(p_s))*3, int(p_l)*3, ' '+sep[1][:-2], len(query_dict[sep[1][:-2]]))\n",
    "                    elif sep[0][:-2] == 'orf1ab_polyprotein_2':\n",
    "                        #print('_2')\n",
    "                        query_dict[sep[1][:-2]] = query_dict[sep[1][:-2]] + cDNA_list[' '+sep[1][:-2]][(int(p_s))*3 : int(p_l)*3]\n",
    "                    else:\n",
    "                        #print('good')\n",
    "                        query_dict[sep[1][:-2]] = cDNA_list[' '+sep[1][:-2]][(int(p_s))*3 : int(p_l)*3]\n",
    "                elif line.split()[1][-1:] == '2':\n",
    "                    if sep[0][:-2] == 'orf1ab_polyprotein_1':\n",
    "                        #print('_1')\n",
    "                        query_dict[sep[1][:-2]] = cDNA_list[' '+sep[1][:-2]][(int(p_s))*3 : int(p_l)*3]\n",
    "                        #print(int(p_s), (int(p_s))*3, int(p_l)*3, ' '+sep[1][:-2], len(query_dict[sep[1][:-2]]))\n",
    "                    elif sep[0][:-2] == 'orf1ab_polyprotein_2':\n",
    "                        #print('_2')\n",
    "                        query_dict[sep[1][:-2]] = query_dict[sep[1][:-2]] + cDNA_list[' '+sep[1][:-2]][(int(p_s))*3 : int(p_l)*3]\n",
    "                    else:\n",
    "                        #print('good')\n",
    "                        query_dict[sep[1][:-2]] = cDNA_list[' '+sep[1][:-2]][(int(p_s))*3 : int(p_l)*3]\n",
    "                elif line.split()[1][-1:] == '3':\n",
    "                    \n",
    "                    if sep[0][:-2] == 'orf1ab_polyprotein_1':\n",
    "                        query_dict[sep[1][:-2]] = cDNA_list[' '+sep[1][:-2]][(int(p_s))*3 : int(p_l)*3]\n",
    "                        #print('_1')\n",
    "                        #print(int(p_s), (int(p_s))*3, int(p_l)*3, ' '+sep[1][:-2], len(query_dict[sep[1][:-2]]))\n",
    "                    elif sep[0][:-2] == 'orf1ab_polyprotein_2':\n",
    "                        #print('_2')\n",
    "                        query_dict[sep[1][:-2]] = query_dict[sep[1][:-2]] + cDNA_list[' '+sep[1][:-2]][(int(p_s))*3 : int(p_l)*3]\n",
    "                    else:\n",
    "                        #print('good')\n",
    "                        query_dict[sep[1][:-2]] = cDNA_list[' '+sep[1][:-2]][(int(p_s))*3 : int(p_l)*3]\n",
    "                temp = (sep[0],sep[1][:-2])\n",
    "    \n",
    "    h = open(\"../data/%s_cDNA_dict.fa\"%query, 'w')\n",
    "    for name in query_dict.keys():\n",
    "        #print(name)\n",
    "        data = \"> %s\\n%s\\n\" %(name, query_dict[name])\n",
    "        h.write(data)\n",
    "    h.close() \n",
    "    f.close()\n",
    "    \n",
    "# Do muscle and get the seuqnces matched nucleotides with the same length  \n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/orf1ab_polyprotein_query_cDNA_dict.fa -out ./data/orf1ab_polyprotein_query_cDNA_dict_muscle.fa \n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/orf1a_polyprotein_query_cDNA_dict.fa -out ./data/orf1a_polyprotein_query_cDNA_dict_muscle.fa  ### ignored\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/surface_glycoprotein_query_cDNA_dict.fa -out ./data/surface_glycoprotein_query_cDNA_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/ORF3a_protein_query_cDNA_dict.fa -out ./data/ORF3a_protein_query_cDNA_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/envelope_protein_query_cDNA_dict.fa -out ./data/envelope_protein_query_cDNA_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/membrane_glycoprotein_query_cDNA_dict.fa -out ./data/membrane_glycoprotein_query_cDNA_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/ORF6_protein_query_cDNA_dict.fa -out ./data/ORF6_protein_query_cDNA_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/ORF7a_protein_query_cDNA_dict.fa -out ./data/ORF7a_protein_query_cDNA_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/ORF7b_query_cDNA_dict.fa -out ./data/ORF7b_query_cDNA_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/ORF8_protein_query_cDNA_dict.fa -out ./data/ORF8_protein_query_cDNA_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/nucleocapsid_phosphoprotein_query_cDNA_dict.fa -out ./data/nucleocapsid_phosphoprotein_query_cDNA_dict_muscle.fa\n",
    "# (terminal) ~/git/BME203-2020$ muscle -in ./data/ORF10_protein_query_cDNA_dict.fa -out ./data/ORF10_protein_query_cDNA_dict_muscle.fa\n",
    "# fasten processing time : -maxiters 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# join each genes into one as one corresponding object.(protein)\n",
    "query_list = ['orf1ab_polyprotein_query', 'surface_glycoprotein_query', 'ORF3a_protein_query', 'envelope_protein_query', 'membrane_glycoprotein_query' ,'ORF6_protein_query' ,'ORF7a_protein_query' ,'ORF7b_query', 'ORF8_protein_query', 'nucleocapsid_phosphoprotein_query', 'ORF10_protein_query']\n",
    "p_set = {}\n",
    "for query in query_list:\n",
    "    dic = read_fasta('../data/%s_protein_dict_muscle.fa'%query)\n",
    "    for i in range(len(dic.keys())):\n",
    "        key = list(dic.keys())[i]\n",
    "        if key in p_set.keys():\n",
    "            p_set[key] = p_set[key] + dic[key]\n",
    "            #print(p_set[key][:25])\n",
    "        else:\n",
    "            p_set[key] = dic[key]\n",
    "\n",
    "\n",
    "\n",
    "h = open(\"../data/SARS_CoV_N_protein_sequence.muscle.fa\", 'w')\n",
    "for name in p_set.keys():\n",
    "    data = \"> %s\\n%s\\n\" %(name, p_set[name])\n",
    "    h.write(data)\n",
    "h.close() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# join each genes into one as one corresponding object.(cDNA except orb1ab)\n",
    "query_list = ['surface_glycoprotein_query', 'ORF3a_protein_query', 'envelope_protein_query', 'membrane_glycoprotein_query' ,'ORF6_protein_query' ,'ORF7a_protein_query' ,'ORF7b_query', 'ORF8_protein_query', 'nucleocapsid_phosphoprotein_query', 'ORF10_protein_query']            \n",
    "c_set = {}\n",
    "for query in query_list:\n",
    "    dic = read_fasta('../data/%s_cDNA_dict_muscle.fa'%query)\n",
    "    for i in range(len(dic.keys())):\n",
    "        key = list(dic.keys())[i]\n",
    "        if key in c_set.keys():\n",
    "            c_set[key] = c_set[key] + dic[key]\n",
    "            #print(c_set[key][:25])\n",
    "        else:\n",
    "            c_set[key] = dic[key]\n",
    "\n",
    "\n",
    "\n",
    "h = open(\"../data/SARS_CoV_N_cDNA_sequence.muscle.fa\", 'w')\n",
    "for name in c_set.keys():\n",
    "    data = \"> %s\\n%s\\n\" %(name, c_set[name])\n",
    "    h.write(data)\n",
    "h.close() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
