{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Task_Midterm_Report. \n",
    "   \n",
    "## Procedures\n",
    "<ol>\n",
    "<li> Collect SARS-CoV2 genes\n",
    "<li> Do Multiple Sequence Alignment & Phylogeny Construction.\n",
    "<li> Calculate the mutation rate estimate of SARS-CoV2\n",
    "<li> Measure the mutation rate estimate for each gene\n",
    "</ol>\n",
    "\n",
    "## Question\n",
    "<ol>\n",
    "<li> Using multiple sequence analysis and phylogenetic tree reconstruction, can you\n",
    "estimate the propagation path of SARS-CoV-2 in the US? Justify your idea with\n",
    "the data.\n",
    "<li> Based on the number of new mutations & the collection time, can you estimate\n",
    "the “mutation rate”? (how long does it take to make a new mutation?)\n",
    "<li> Is there any specific gene having higher rate of mutation? Can you explain why?\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The function to read sequences from a FASTA file\n",
    "def read_fasta(tmp):\n",
    "    r = dict()\n",
    "    fi = open(tmp, 'r')\n",
    "    for line in fi:\n",
    "        if line.startswith('>'):\n",
    "            h_g = line.strip().lstrip('>')\n",
    "            r[h_g] = ''\n",
    "        else:\n",
    "            r[h_g] += line.strip().replace(' ', '')\n",
    "    fi.close()\n",
    "    return r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1-(1) : Read genomes\n",
    "\n",
    "SARS_CoV2_genomes = '2019nCoV_genomes.2020_04_14.clean.fa'\n",
    "genome_COVID = read_fasta(SARS_CoV2_genomes)\n",
    "CoV2_rep_idx = '../data/GCF_009858895.2_ASM985889v3_genomic.gff'\n",
    "ref_genome = '../data/GCF_009858895.2_ASM985889v3_genomic.fna'\n",
    "genome_COVID_ref = read_fasta(ref_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1-(2) : get index\n",
    "\n",
    "def get_idx(t, p):\n",
    "    fi = open(t, 'r')\n",
    "    for line in fi.readlines():\n",
    "        if \"gene\" in line and \"Name={}\".format(p) in line:\n",
    "            idx_s = int(line.split()[3])-1\n",
    "            idx_e = int(line.split()[4])-1\n",
    "            idx = (idx_s, idx_e)\n",
    "    return idx\n",
    "N_idx = get_idx(CoV2_rep_idx, 'N')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1-(3) : Get the cDNA seq of N protein in SARS-CoV2\n",
    "\n",
    "for seq in genome_COVID_ref.values():\n",
    "    ref_N_cDNA = seq[N_idx[0]:N_idx[1]+1]\n",
    "ref_Np_cDNA = {'N_prot_cDNA' : [ref_N_cDNA]}\n",
    "def w_seq(fi, gc):\n",
    "    f = open(fi, 'w')\n",
    "    for g in list(gc.keys()):\n",
    "        if len(gc[g]) == 1:\n",
    "            f.write(\">{}\\n\".format(g))\n",
    "            f.write(\"{}\\n\".format(gc[g][0]))\n",
    "        else:\n",
    "            for i in range(len(gc[g])):\n",
    "                f.write(\">{} #{}\\n\".format(g, i+1))\n",
    "                f.write(\"\\n{}\".format(gc[g][i]))\n",
    "    f.close()\n",
    "w_seq(\"ref_Np_cDNA.fa\", ref_Np_cDNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1-(4) : Blast about SARS-CoV2 N gene and Make .fa file about N gene sequence data of each 2019nCoV_genomes.2020_04_14.clean.fa\n",
    "\n",
    "'''\n",
    "(base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ makeblastdb -dbtype nucl -in 2019nCoV_genomes.2020_04_14.clean.fa -out 2019nCoV_genomes.2020_04_14.clean\n",
    "(base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ blastn -db 2019nCoV_genomes.2020_04_14.clean -query ref_Np_cDNA.fa -outfmt 6 -out 2019nCoV_genomes_N.2020_04_14_blastn.out -max_target_seqs 746\n",
    "'''\n",
    "\n",
    "cD_set = open(\"SARS_CoV2_N_set.fa\", 'w')\n",
    "gf = open('2019nCoV_genomes_N.2020_04_14_blastn.out', 'r')\n",
    "\n",
    "val_list = []\n",
    "lines = gf.readlines()\n",
    "\n",
    "for line in lines:\n",
    "    for val in list(genome_COVID.keys()):\n",
    "        if line.split()[1] in val:\n",
    "            val_list.append(line.split()[1])\n",
    "\n",
    "val_set = set(val_list)\n",
    "\n",
    "cD_dic = {val : [] for val in val_set}\n",
    "\n",
    "for line in lines:\n",
    "    cD_dic[line.split()[1]].append(line.split()[8])\n",
    "    cD_dic[line.split()[1]].append(line.split()[9])\n",
    "          \n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    if len(cD_dic[val]) > 2:\n",
    "        max_idx = max(cD_dic[val])\n",
    "        min_idx = min(cD_dic[val])\n",
    "        cD_dic[val].clear()\n",
    "        cD_dic[val].append(min_idx)\n",
    "        cD_dic[val].append(max_idx)\n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    for key in genome_COVID.keys():\n",
    "        if val in key:\n",
    "            idx_1 = int(cD_dic[val][0])-1\n",
    "            idx_2 = int(cD_dic[val][1])\n",
    "            sen = \">%s\\n%s\\n\" % (val, genome_COVID[key][idx_1:idx_2])\n",
    "            cD_set.write(sen)\n",
    "\n",
    "\n",
    "gf.close()\n",
    "cD_set.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1-(5) : Do multiple alignment and construct phylogenetic tree about N gene by using Mega X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# But, Phylogenetic tree from N gene has low accuracy because N gene is just only one gene of some genes in SARS-CoV2\n",
    "# Thus, I'll select the nine genes(S, ORF1ab, ORF8, E, M, ORF3a, ORF6, ORF7a) and reconstruct phylogenetic tree from them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1-(6) :  reference sequence about S & ORF1ab & ORF8 & E & M & ORF3a & ORF6 & ORF7a\n",
    "\n",
    "S_idx = get_idx(CoV2_rep_idx, 'S')\n",
    "ORF8_idx = get_idx(CoV2_rep_idx, 'YP_009724396.1')\n",
    "E_idx = get_idx(CoV2_rep_idx, 'E')\n",
    "M_idx = get_idx(CoV2_rep_idx, 'M')\n",
    "ORF1ab_idx = get_idx(CoV2_rep_idx, 'orf1ab')\n",
    "ORF3a_idx = get_idx(CoV2_rep_idx, 'ORF3a')\n",
    "M_idx = get_idx(CoV2_rep_idx, 'M')\n",
    "ORF6_idx = get_idx(CoV2_rep_idx, 'ORF6')\n",
    "ORF7a_idx = get_idx(CoV2_rep_idx, 'ORF7a')\n",
    "\n",
    "\n",
    "for seq in genome_COVID_ref.values():\n",
    "    ref_S_cDNA = seq[S_idx[0]:S_idx[1]+1]\n",
    "ref_Sp_cDNA = {'S_prot_cDNA' : [ref_S_cDNA]}\n",
    "\n",
    "for seq in genome_COVID_ref.values():\n",
    "    ref_ORF8_cDNA = seq[ORF8_idx[0]:ORF8_idx[1]+1]\n",
    "ref_ORF8p_cDNA = {'ORF8_prot_cDNA' : [ref_ORF8_cDNA]}\n",
    "\n",
    "for seq in genome_COVID_ref.values():\n",
    "    ref_E_cDNA = seq[E_idx[0]:E_idx[1]+1]\n",
    "ref_Ep_cDNA = {'E_prot_cDNA' : [ref_E_cDNA]}\n",
    "\n",
    "for seq in genome_COVID_ref.values():\n",
    "    ref_ORF1ab_cDNA = seq[ORF1ab_idx[0]:ORF1ab_idx[1]+1]\n",
    "ref_ORF1abp_cDNA = {'ORF1ab_prot_cDNA' : [ref_ORF1ab_cDNA]}\n",
    "\n",
    "for seq in genome_COVID_ref.values():\n",
    "    ref_ORF3a_cDNA = seq[ORF3a_idx[0]:ORF3a_idx[1]+1]\n",
    "ref_ORF3ap_cDNA = {'ORF3a_prot_cDNA' : [ref_ORF3a_cDNA]}\n",
    "\n",
    "for seq in genome_COVID_ref.values():\n",
    "    ref_M_cDNA = seq[M_idx[0]:M_idx[1]+1]\n",
    "ref_Mp_cDNA = {'M_prot_cDNA' : [ref_M_cDNA]}\n",
    "\n",
    "for seq in genome_COVID_ref.values():\n",
    "    ref_ORF6_cDNA = seq[ORF6_idx[0]:ORF6_idx[1]+1]\n",
    "ref_ORF6p_cDNA = {'ORF6_prot_cDNA' : [ref_ORF6_cDNA]}\n",
    "\n",
    "for seq in genome_COVID_ref.values():\n",
    "    ref_ORF7a_cDNA = seq[ORF7a_idx[0]:ORF7a_idx[1]+1]\n",
    "ref_ORF7ap_cDNA = {'ORF7a_prot_cDNA' : [ref_ORF7a_cDNA]}\n",
    "\n",
    "\n",
    "\n",
    "w_seq(\"ref_Sp_cDNA.fa\", ref_Sp_cDNA)\n",
    "w_seq(\"ref_ORF8p_cDNA.fa\", ref_ORF8p_cDNA)\n",
    "w_seq(\"ref_Ep_cDNA.fa\", ref_Ep_cDNA)\n",
    "w_seq(\"ref_ORF1abp_cDNA.fa\", ref_ORF1abp_cDNA)\n",
    "w_seq(\"ref_ORF3ap_cDNA.fa\", ref_ORF3ap_cDNA)\n",
    "w_seq(\"ref_Mp_cDNA.fa\", ref_Mp_cDNA)\n",
    "w_seq(\"ref_ORF6p_cDNA.fa\", ref_ORF6p_cDNA)\n",
    "w_seq(\"ref_ORF7ap_cDNA.fa\", ref_ORF7ap_cDNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1-(7) : Blast about nine genes except for N gene\n",
    "\n",
    "'''\n",
    "(base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ blastn -db 2019nCoV_genomes.2020_04_14.clean -query ref_Sp_cDNA.fa -outfmt 6 -out 2019nCoV_genomes_S.2020_04_14_blastn.out -max_target_seqs 746\n",
    "(base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ blastn -db 2019nCoV_genomes.2020_04_14.clean -query ref_ORF8p_cDNA.fa -outfmt 6 -out 2019nCoV_genomes_ORF8.2020_04_14_blastn.out -max_target_seqs 746\n",
    "(base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ blastn -db 2019nCoV_genomes.2020_04_14.clean -query ref_Ep_cDNA.fa -outfmt 6 -out 2019nCoV_genomes_E.2020_04_14_blastn.out -max_target_seqs 746\n",
    "(base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ blastn -db 2019nCoV_genomes.2020_04_14.clean -query ref_ORF1abp_cDNA.fa -outfmt 6 -out 2019nCoV_genomes_ORF1ab.2020_04_14_blastn.out -max_target_seqs 746\n",
    "(base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ blastn -db 2019nCoV_genomes.2020_04_14.clean -query ref_ORF3ap_cDNA.fa -outfmt 6 -out 2019nCoV_genomes_ORF3a.2020_04_14_blastn.out -max_target_seqs 746\n",
    "(base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ blastn -db 2019nCoV_genomes.2020_04_14.clean -query ref_Mp_cDNA.fa -outfmt 6 -out 2019nCoV_genomes_M.2020_04_14_blastn.out -max_target_seqs 746\n",
    "(base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ blastn -db 2019nCoV_genomes.2020_04_14.clean -query ref_ORF6p_cDNA.fa -outfmt 6 -out 2019nCoV_genomes_ORF6.2020_04_14_blastn.out -max_target_seqs 746\n",
    "(base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ blastn -db 2019nCoV_genomes.2020_04_14.clean -query ref_ORF7ap_cDNA.fa -outfmt 6 -out 2019nCoV_genomes_ORF7a.2020_04_14_blastn.out -max_target_seqs 746\n",
    "'''\n",
    "\n",
    "# 1-(8) : Make .fa file about sequence data(from nine genes) of each 2019nCoV_genomes.2020_04_14.clean.fa\n",
    "# > S gene\n",
    "\n",
    "cD_set_S = open(\"SARS_CoV2_S_set.fa\", 'w')\n",
    "gf_S = open('2019nCoV_genomes_S.2020_04_14_blastn.out', 'r')\n",
    "\n",
    "val_list = []\n",
    "lines = gf_S.readlines()\n",
    "\n",
    "for line in lines:\n",
    "    for val in list(genome_COVID.keys()):\n",
    "        if line.split()[1] in val:\n",
    "            val_list.append(line.split()[1])\n",
    "\n",
    "val_set = set(val_list)\n",
    "\n",
    "cD_dic = {val : [] for val in val_set}\n",
    "\n",
    "for line in lines:\n",
    "    cD_dic[line.split()[1]].append(line.split()[8])\n",
    "    cD_dic[line.split()[1]].append(line.split()[9])\n",
    "          \n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    if len(cD_dic[val]) > 2:\n",
    "        max_idx = max(cD_dic[val])\n",
    "        min_idx = min(cD_dic[val])\n",
    "        cD_dic[val].clear()\n",
    "        cD_dic[val].append(min_idx)\n",
    "        cD_dic[val].append(max_idx)\n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    for key in genome_COVID.keys():\n",
    "        if val in key:\n",
    "            idx_1 = int(cD_dic[val][0])-1\n",
    "            idx_2 = int(cD_dic[val][1])\n",
    "            sen = \">%s\\n%s\\n\" % (val, genome_COVID[key][idx_1:idx_2])\n",
    "            cD_set_S.write(sen)\n",
    "\n",
    "gf_S.close()\n",
    "cD_set_S.close()   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# > E gene & ORF8 & ORF1ab\n",
    "\n",
    "cD_set_E = open(\"SARS_CoV2_E_set.fa\", 'w')\n",
    "gf_E = open('2019nCoV_genomes_E.2020_04_14_blastn.out', 'r')\n",
    "\n",
    "val_list = []\n",
    "lines = gf_E.readlines()\n",
    "\n",
    "for line in lines:\n",
    "    for val in list(genome_COVID.keys()):\n",
    "        if line.split()[1] in val:\n",
    "            val_list.append(line.split()[1])\n",
    "\n",
    "val_set = set(val_list)\n",
    "\n",
    "cD_dic = {val : [] for val in val_set}\n",
    "\n",
    "for line in lines:\n",
    "    cD_dic[line.split()[1]].append(line.split()[8])\n",
    "    cD_dic[line.split()[1]].append(line.split()[9])\n",
    "          \n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    if len(cD_dic[val]) > 2:\n",
    "        max_idx = max(cD_dic[val])\n",
    "        min_idx = min(cD_dic[val])\n",
    "        cD_dic[val].clear()\n",
    "        cD_dic[val].append(min_idx)\n",
    "        cD_dic[val].append(max_idx)\n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    for key in genome_COVID.keys():\n",
    "        if val in key:\n",
    "            idx_1 = int(cD_dic[val][0])-1\n",
    "            idx_2 = int(cD_dic[val][1])\n",
    "            sen = \">%s\\n%s\\n\" % (val, genome_COVID[key][idx_1:idx_2])\n",
    "            cD_set_E.write(sen)\n",
    "\n",
    "gf_E.close()\n",
    "cD_set_E.close()\n",
    "\n",
    "cD_set_ORF8 = open(\"SARS_CoV2_ORF8_set.fa\", 'w')\n",
    "gf_ORF8 = open('2019nCoV_genomes_ORF8.2020_04_14_blastn.out', 'r')\n",
    "\n",
    "val_list = []\n",
    "lines = gf_ORF8.readlines()\n",
    "\n",
    "for line in lines:\n",
    "    for val in list(genome_COVID.keys()):\n",
    "        if line.split()[1] in val:\n",
    "            val_list.append(line.split()[1])\n",
    "\n",
    "val_set = set(val_list)\n",
    "\n",
    "cD_dic = {val : [] for val in val_set}\n",
    "\n",
    "for line in lines:\n",
    "    cD_dic[line.split()[1]].append(line.split()[8])\n",
    "    cD_dic[line.split()[1]].append(line.split()[9])\n",
    "          \n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    if len(cD_dic[val]) > 2:\n",
    "        max_idx = max(cD_dic[val])\n",
    "        min_idx = min(cD_dic[val])\n",
    "        cD_dic[val].clear()\n",
    "        cD_dic[val].append(min_idx)\n",
    "        cD_dic[val].append(max_idx)\n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    for key in genome_COVID.keys():\n",
    "        if val in key:\n",
    "            idx_1 = int(cD_dic[val][0])-1\n",
    "            idx_2 = int(cD_dic[val][1])\n",
    "            sen = \">%s\\n%s\\n\" % (val, genome_COVID[key][idx_1:idx_2])\n",
    "            cD_set_ORF8.write(sen)\n",
    "\n",
    "gf_ORF8.close()\n",
    "cD_set_ORF8.close()\n",
    "\n",
    "cD_set_ORF1ab = open(\"SARS_CoV2_ORF1ab_d_set.fa\", 'w')\n",
    "gf_ORF1ab = open('2019nCoV_genomes_ORF1ab.2020_04_14_blastn.out', 'r')\n",
    "\n",
    "val_list = []\n",
    "lines = gf_ORF1ab.readlines()\n",
    "\n",
    "for line in lines:\n",
    "    for val in list(genome_COVID.keys()):\n",
    "        if line.split()[1] in val:\n",
    "            val_list.append(line.split()[1])\n",
    "\n",
    "val_set = set(val_list)\n",
    "\n",
    "cD_dic = {val : [] for val in val_set}\n",
    "\n",
    "for line in lines:\n",
    "    cD_dic[line.split()[1]].append(line.split()[8])\n",
    "    cD_dic[line.split()[1]].append(line.split()[9])\n",
    "          \n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    if len(cD_dic[val]) > 2:\n",
    "        max_idx = max(cD_dic[val])\n",
    "        min_idx = min(cD_dic[val])\n",
    "        cD_dic[val].clear()\n",
    "        cD_dic[val].append(min_idx)\n",
    "        cD_dic[val].append(max_idx)\n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    for key in genome_COVID.keys():\n",
    "        if val in key:\n",
    "            idx_1 = int(cD_dic[val][0])-1\n",
    "            idx_2 = int(cD_dic[val][1])\n",
    "            sen = \">%s\\n%s\\n\" % (val, genome_COVID[val][idx_1:idx_2])\n",
    "            cD_set_ORF1ab.write(sen)         \n",
    "            \n",
    "        \n",
    "        \n",
    "\n",
    "gf_ORF1ab.close()\n",
    "cD_set_ORF1ab.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# > ORF3a & M gene\n",
    "\n",
    "cD_set_ORF3a = open(\"SARS_CoV2_ORF3a_set.fa\", 'w')\n",
    "gf_ORF3a = open('2019nCoV_genomes_ORF3a.2020_04_14_blastn.out', 'r')\n",
    "\n",
    "val_list = []\n",
    "lines = gf_ORF3a.readlines()\n",
    "\n",
    "for line in lines:\n",
    "    for val in list(genome_COVID.keys()):\n",
    "        if line.split()[1] in val:\n",
    "            val_list.append(line.split()[1])\n",
    "\n",
    "val_set = set(val_list)\n",
    "\n",
    "cD_dic = {val : [] for val in val_set}\n",
    "\n",
    "for line in lines:\n",
    "    cD_dic[line.split()[1]].append(line.split()[8])\n",
    "    cD_dic[line.split()[1]].append(line.split()[9])\n",
    "          \n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    if len(cD_dic[val]) > 2:\n",
    "        max_idx = max(cD_dic[val])\n",
    "        min_idx = min(cD_dic[val])\n",
    "        cD_dic[val].clear()\n",
    "        cD_dic[val].append(min_idx)\n",
    "        cD_dic[val].append(max_idx)\n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    for key in genome_COVID.keys():\n",
    "        if val in key:\n",
    "            idx_1 = int(cD_dic[val][0])-1\n",
    "            idx_2 = int(cD_dic[val][1])\n",
    "            sen = \">%s\\n%s\\n\" % (val, genome_COVID[key][idx_1:idx_2])\n",
    "            cD_set_ORF3a.write(sen)\n",
    "\n",
    "gf_ORF3a.close()\n",
    "cD_set_ORF3a.close()\n",
    "\n",
    "cD_set_M = open(\"SARS_CoV2_M_set.fa\", 'w')\n",
    "gf_M = open('2019nCoV_genomes_M.2020_04_14_blastn.out', 'r')\n",
    "\n",
    "val_list = []\n",
    "lines = gf_M.readlines()\n",
    "\n",
    "for line in lines:\n",
    "    for val in list(genome_COVID.keys()):\n",
    "        if line.split()[1] in val:\n",
    "            val_list.append(line.split()[1])\n",
    "\n",
    "val_set = set(val_list)\n",
    "\n",
    "cD_dic = {val : [] for val in val_set}\n",
    "\n",
    "for line in lines:\n",
    "    cD_dic[line.split()[1]].append(line.split()[8])\n",
    "    cD_dic[line.split()[1]].append(line.split()[9])\n",
    "          \n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    if len(cD_dic[val]) > 2:\n",
    "        max_idx = max(cD_dic[val])\n",
    "        min_idx = min(cD_dic[val])\n",
    "        cD_dic[val].clear()\n",
    "        cD_dic[val].append(min_idx)\n",
    "        cD_dic[val].append(max_idx)\n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    for key in genome_COVID.keys():\n",
    "        if val in key:\n",
    "            idx_1 = int(cD_dic[val][0])-1\n",
    "            idx_2 = int(cD_dic[val][1])\n",
    "            sen = \">%s\\n%s\\n\" % (val, genome_COVID[key][idx_1:idx_2])\n",
    "            cD_set_M.write(sen)\n",
    "\n",
    "gf_M.close()\n",
    "cD_set_M.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# > ORF6 & ORF7a\n",
    "\n",
    "cD_set_ORF6 = open(\"SARS_CoV2_ORF6_set.fa\", 'w')\n",
    "gf_ORF6 = open('2019nCoV_genomes_ORF6.2020_04_14_blastn.out', 'r')\n",
    "\n",
    "val_list = []\n",
    "lines = gf_ORF6.readlines()\n",
    "\n",
    "for line in lines:\n",
    "    for val in list(genome_COVID.keys()):\n",
    "        if line.split()[1] in val:\n",
    "            val_list.append(line.split()[1])\n",
    "\n",
    "val_set = set(val_list)\n",
    "\n",
    "cD_dic = {val : [] for val in val_set}\n",
    "\n",
    "for line in lines:\n",
    "    cD_dic[line.split()[1]].append(line.split()[8])\n",
    "    cD_dic[line.split()[1]].append(line.split()[9])\n",
    "          \n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    if len(cD_dic[val]) > 2:\n",
    "        max_idx = max(cD_dic[val])\n",
    "        min_idx = min(cD_dic[val])\n",
    "        cD_dic[val].clear()\n",
    "        cD_dic[val].append(min_idx)\n",
    "        cD_dic[val].append(max_idx)\n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    for key in genome_COVID.keys():\n",
    "        if val in key:\n",
    "            idx_1 = int(cD_dic[val][0])-1\n",
    "            idx_2 = int(cD_dic[val][1])\n",
    "            sen = \">%s\\n%s\\n\" % (val, genome_COVID[key][idx_1:idx_2])\n",
    "            cD_set_ORF6.write(sen)\n",
    "\n",
    "gf_ORF6.close()\n",
    "cD_set_ORF6.close()\n",
    "\n",
    "cD_set_ORF7a = open(\"SARS_CoV2_ORF7a_set.fa\", 'w')\n",
    "gf_ORF7a = open('2019nCoV_genomes_ORF7a.2020_04_14_blastn.out', 'r')\n",
    "\n",
    "val_list = []\n",
    "lines = gf_ORF7a.readlines()\n",
    "\n",
    "for line in lines:\n",
    "    for val in list(genome_COVID.keys()):\n",
    "        if line.split()[1] in val:\n",
    "            val_list.append(line.split()[1])\n",
    "\n",
    "val_set = set(val_list)\n",
    "\n",
    "cD_dic = {val : [] for val in val_set}\n",
    "\n",
    "for line in lines:\n",
    "    cD_dic[line.split()[1]].append(line.split()[8])\n",
    "    cD_dic[line.split()[1]].append(line.split()[9])\n",
    "          \n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    if len(cD_dic[val]) > 2:\n",
    "        max_idx = max(cD_dic[val])\n",
    "        min_idx = min(cD_dic[val])\n",
    "        cD_dic[val].clear()\n",
    "        cD_dic[val].append(min_idx)\n",
    "        cD_dic[val].append(max_idx)\n",
    "        \n",
    "for val in cD_dic.keys():\n",
    "    for key in genome_COVID.keys():\n",
    "        if val in key:\n",
    "            idx_1 = int(cD_dic[val][0])-1\n",
    "            idx_2 = int(cD_dic[val][1])\n",
    "            sen = \">%s\\n%s\\n\" % (val, genome_COVID[key][idx_1:idx_2])\n",
    "            cD_set_ORF7a.write(sen)\n",
    "\n",
    "gf_ORF7a.close()\n",
    "cD_set_ORF7a.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# > Insert missing ORF1ab sequence data -> (But, I don't know why some data is missing)\n",
    "\n",
    "cD_ORF1ab = \"SARS_CoV2_ORF1ab_d_set.fa\"\n",
    "ORF1ab = read_fasta(cD_ORF1ab)\n",
    "\n",
    "ORF1ab_dic = {}\n",
    "\n",
    "for gORF1ab in list(ORF1ab.keys()):\n",
    "    for key in genome_COVID.keys():\n",
    "        if gORF1ab in key:\n",
    "            ORF1ab_dic[key] = ORF1ab[gORF1ab]\n",
    "            \n",
    "for n_key in ORF1ab_dic.keys():\n",
    "    if n_key == 'MT292579|Spain|2020-03-09':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT292579|Spain|2020-03-09'][211:21501]\n",
    "    elif n_key == 'MT326115|USA|2020-03-20':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326115|USA|2020-03-20'][0:21249]\n",
    "    elif n_key == 'MT322399|USA_VA|2020-03-00':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT322399|USA_VA|2020-03-00'][226:21516]\n",
    "    elif n_key == 'MT198651|Spain_Valencia|2020-03-04':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT198651|Spain_Valencia|2020-03-04'][211:21501]\n",
    "    elif n_key == 'MT259272|USA_WA|2020-03-13':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT259272|USA_WA|2020-03-13'][0:21136]\n",
    "    elif n_key == 'MT326183|USA|2020-03-17':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326183|USA|2020-03-17'][119:21407]\n",
    "    elif n_key == 'MT326038|USA|2020-03-24':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326038|USA|2020-03-24'][0:21184]\n",
    "    elif n_key == 'MT326108|USA|2020-03-22':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326108|USA|2020-03-22'][0:21251]\n",
    "    elif n_key == 'MT308695|USA_MI|2020-03-10':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT308695|USA_MI|2020-03-10'][216:21506]\n",
    "    elif n_key == 'MT263407|USA_WA|2020-03-22':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT263407|USA_WA|2020-03-22'][0:21279]\n",
    "    elif n_key == 'MT326142|USA|2020-03-19':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326142|USA|2020-03-19'][0:21053]\n",
    "    elif n_key == 'MT259242|USA_WA|2020-03-14':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT259242|USA_WA|2020-03-14'][0:20911]\n",
    "    elif n_key == 'MT326131|USA|2020-03-20':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326131|USA|2020-03-20'][0:21098]\n",
    "    elif n_key == 'MT246458|USA_WA|2020-03-12':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT246458|USA_WA|2020-03-12'][139:21429]\n",
    "    elif n_key == 'MT259238|USA_WA|2020-03-16':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT259238|USA_WA|2020-03-16'][0:20813]\n",
    "    elif n_key == 'MT263397|USA_WA|2020-03-24':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT263397|USA_WA|2020-03-24'][0:21254]\n",
    "    elif n_key == 'MT263426|USA_WA|2020-03-24':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT263426|USA_WA|2020-03-24'][154:21444]\n",
    "    elif n_key == 'MT326146|USA|2020-03-21':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326146|USA|2020-03-21'][0:21190]\n",
    "    elif n_key == 'MT263453|USA_WA|2020-03-24':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT263453|USA_WA|2020-03-24'][139:21429]\n",
    "    elif n_key == 'MT322400|USA_VA|2020-03-00':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT322400|USA_VA|2020-03-00'][249:21539]\n",
    "    elif n_key == 'MT259284|USA_WA|2020-03-15':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT259284|USA_WA|2020-03-15'][0:21156]\n",
    "    elif n_key == 'MT308692|USA_MI|2020-03-09':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT308692|USA_MI|2020-03-09'][218:21508]\n",
    "    elif n_key == 'MT326102|USA|2020-03-22':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326102|USA|2020-03-22'][19:21309]\n",
    "    elif n_key == 'MT308696|USA_MI|2020-03-09':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT308696|USA_MI|2020-03-09'][216:21506]\n",
    "    elif n_key == 'MT246483|USA_WA|2020-03-15':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT246483|USA_WA|2020-03-15'][196:21429]\n",
    "    elif n_key == 'MT263389|USA_WA|2020-03-23':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT263389|USA_WA|2020-03-23'][0:21106]\n",
    "    elif n_key == 'MT293203|USA_WA|2020-03-30':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT293203|USA_WA|2020-03-30'][0:21251]\n",
    "    elif n_key == 'MT308700|USA_MI|2020-03-12':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT308700|USA_MI|2020-03-12'][216:21506]\n",
    "    elif n_key == 'MT246463|USA_WA|2020-03-13':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT246463|USA_WA|2020-03-13'][139:21429]\n",
    "    elif n_key == 'MT326145|USA|2020-03-21':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326145|USA|2020-03-21'][0:21060]\n",
    "    elif n_key == 'MT259283|USA_WA|2020-03-15':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT259283|USA_WA|2020-03-15'][0:20939]\n",
    "    elif n_key == 'MT308693|USA_MI|2020-03-10':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT308693|USA_MI|2020-03-10'][216:21506]\n",
    "    elif n_key == 'MT263461|USA_WA|2020-03-22':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT263461|USA_WA|2020-03-22'][235:21481]\n",
    "    elif n_key == 'MT246465|USA_WA|2020-03-13':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT246465|USA_WA|2020-03-13'][241:21433]\n",
    "    elif n_key == 'MT259270|USA_WA|2020-03-16':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT259270|USA_WA|2020-03-16'][0:21088]\n",
    "    elif n_key == 'MT256918|Spain_Valencia|2020-03-06':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT256918|Spain_Valencia|2020-03-06'][211:21501]\n",
    "    elif n_key == 'MT326101|USA|2020-03-22':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326101|USA|2020-03-22'][250:21407]\n",
    "    elif n_key == 'MT256917|Spain_Valencia|2020-03-02':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT256917|Spain_Valencia|2020-03-02'][211:21501]\n",
    "    elif n_key == 'MT308699|USA_MI|2020-03-12':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT308699|USA_MI|2020-03-12'][210:21500]\n",
    "    elif n_key == 'MT326186|USA|2020-03-17':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326186|USA|2020-03-17'][0:21251]\n",
    "    elif n_key == 'MT259255|USA_WA|2020-03-16':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT259255|USA_WA|2020-03-16'][0:21093]\n",
    "    elif n_key == 'MT326032|USA|2020-03-23':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326032|USA|2020-03-23'][274:21506]\n",
    "    elif n_key == 'MT326126|USA|2020-03-20':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326126|USA|2020-03-20'][300:21508]\n",
    "    elif n_key == 'MT292580|Spain|2020-02-27':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT292580|Spain|2020-02-27'][211:21501]\n",
    "    elif n_key == 'MT263427|USA_WA|2020-03-24':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT263427|USA_WA|2020-03-24'][320:21462]\n",
    "    elif n_key == 'MT292582|Spain|2020-03-06':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT292582|Spain|2020-03-06'][211:21501]\n",
    "    elif n_key == 'MT326176|USA|2020-03-19':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326176|USA|2020-03-19'][0:20992]\n",
    "    elif n_key == 'MT326057|USA|2020-03-21':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326057|USA|2020-03-21'][0:21177]\n",
    "    elif n_key == 'MT308694|USA_MI|2020-03-11':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT308694|USA_MI|2020-03-11'][216:21506]\n",
    "    elif n_key == 'MT263460|USA_WA|2020-03-24':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT263460|USA_WA|2020-03-24'][124:21407]\n",
    "    elif n_key == 'MT263393|USA_WA|2020-03-23':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT263393|USA_WA|2020-03-23'][0:21182]\n",
    "    elif n_key == 'MT263387|USA_WA|2020-03-23':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT263387|USA_WA|2020-03-23'][203:21493]\n",
    "    elif n_key == 'MT308697|USA_MI|2020-03-11':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT308697|USA_MI|2020-03-11'][218:21508]\n",
    "    elif n_key == 'MT326080|USA|2020-03-24':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326080|USA|2020-03-24'][43:21136]\n",
    "    elif n_key == 'LR757997|China_Wuhan|2019-12-31':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['LR757997|China_Wuhan|2019-12-31'][229:21560]\n",
    "    elif n_key == 'MT259259|USA_WA|2020-03-16':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT259259|USA_WA|2020-03-16'][0:21037]\n",
    "    elif n_key == 'MT292581|Spain|2020-02-27':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT292581|Spain|2020-02-27'][211:21501]\n",
    "    elif n_key == 'MT263466|USA_WA|2020-03-23':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT263466|USA_WA|2020-03-23'][99:21376]\n",
    "    elif n_key == 'MT326179|USA|2020-03-19':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326179|USA|2020-03-19'][99:21389]\n",
    "    elif n_key == 'MT326182|USA|2020-03-13':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326182|USA|2020-03-13'][51:21341]\n",
    "    elif n_key == 'MT263401|USA_WA|2020-03-23':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT263401|USA_WA|2020-03-23'][0:21175]\n",
    "    elif n_key == 'MT263409|USA_WA|2020-03-22':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT263409|USA_WA|2020-03-22'][0:21190]\n",
    "    elif n_key == 'MT293180|USA_WA|2020-03-28':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT293180|USA_WA|2020-03-28'][43:21293]\n",
    "    elif n_key == 'MT308698|USA_MI|2020-03-12':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT308698|USA_MI|2020-03-12'][216:21506]\n",
    "    elif n_key == 'MT326073|USA|2020-03-21':\n",
    "        ORF1ab_dic[n_key] = genome_COVID['MT326073|USA|2020-03-21'][0:21192]\n",
    "\n",
    "\n",
    "cD_r_ORFab = open('SARS_CoV2_ORF1ab_set.fa', 'w')\n",
    "\n",
    "for o in list(ORF1ab_dic.keys()):\n",
    "    cD_r_ORFab.write(\">{}\\n\".format(o))\n",
    "    cD_r_ORFab.write(\"{}\\n\".format(ORF1ab_dic[o]))    \n",
    "    \n",
    "cD_r_ORFab.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1-(9) : Do multiple Alignment about nine genes and merge all aligned data from each gene into one .fa file\n",
    "\n",
    "# (base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ muscle -in SARS_CoV2_ORF1ab_set.fa -out SARS_CoV2_ORF1ab_mul_al.fa\n",
    "# (base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ muscle -in SARS_CoV2_ORF8_set.fa -out SARS_CoV2_ORF8_mul_al.fa\n",
    "# (base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ muscle -in SARS_CoV2_N_set.fa -out SARS_CoV2_N_mul_al.fa\n",
    "# (base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ muscle -in SARS_CoV2_E_set.fa -out SARS_CoV2_E_mul_al.fa\n",
    "# (base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ muscle -in SARS_CoV2_S_set.fa -out SARS_CoV2_S_mul_al.fa\n",
    "# (base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ muscle -in SARS_CoV2_ORF3a_set.fa -out SARS_CoV2_ORF3a_mul_al.fa\n",
    "# (base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ muscle -in SARS_CoV2_M_set.fa -out SARS_CoV2_M_mul_al.fa\n",
    "# (base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ muscle -in SARS_CoV2_ORF6_set.fa -out SARS_CoV2_ORF6_mul_al.fa\n",
    "# (base) mirna@DESKTOP-BN3PD9C:~/git/BME203-2020/05_Report$ muscle -in SARS_CoV2_ORF7a_set.fa -out SARS_CoV2_ORF7a_mul_al.fa\n",
    "\n",
    "alig_N = \"SARS_CoV2_N_mul_al.fa\"\n",
    "aN = read_fasta(alig_N)\n",
    "alig_S = \"SARS_CoV2_S_mul_al.fa\"\n",
    "aS = read_fasta(alig_S)\n",
    "alig_E = \"SARS_CoV2_E_mul_al.fa\"\n",
    "aE = read_fasta(alig_E)\n",
    "alig_ORF8 = \"SARS_CoV2_ORF8_mul_al.fa\"\n",
    "aORF8 = read_fasta(alig_ORF8)\n",
    "alig_ORF1ab = \"SARS_CoV2_ORF1ab_mul_al.fa\"\n",
    "aORF1ab = read_fasta(alig_ORF1ab)\n",
    "alig_ORF3a = \"SARS_CoV2_ORF3a_mul_al.fa\"\n",
    "aORF3a = read_fasta(alig_ORF3a)\n",
    "alig_M = \"SARS_CoV2_M_mul_al.fa\"\n",
    "aM = read_fasta(alig_M)\n",
    "alig_ORF6 = \"SARS_CoV2_ORF6_mul_al.fa\"\n",
    "aORF6 = read_fasta(alig_ORF6)\n",
    "alig_ORF7a = \"SARS_CoV2_ORF7a_mul_al.fa\"\n",
    "aORF7a = read_fasta(alig_ORF7a)\n",
    "\n",
    "cD_treeg_dic = {}\n",
    "\n",
    "for gORF1ab in list(aORF1ab.keys()):\n",
    "    if gORF1ab == 'NC_045512|China|2019-12-00':\n",
    "        cD_treeg_dic[gORF1ab] = aORF1ab[gORF1ab]\n",
    "\n",
    "for gORF1ab in list(aORF1ab.keys()):\n",
    "    if gORF1ab != 'NC_045512|China|2019-12-00':\n",
    "        cD_treeg_dic[gORF1ab] = aORF1ab[gORF1ab]\n",
    "    \n",
    "for gS in list(aS.keys()):\n",
    "    for val in cD_treeg_dic.keys():\n",
    "        if gS == val:\n",
    "            cD_treeg_dic[val] += aS[gS]\n",
    "            \n",
    "for gORF3a in list(aORF3a.keys()):\n",
    "    for val in cD_treeg_dic.keys():\n",
    "        if gORF3a == val:\n",
    "            cD_treeg_dic[val] += aORF3a[gORF3a]            \n",
    "            \n",
    "for gE in list(aE.keys()):\n",
    "    for val in cD_treeg_dic.keys():\n",
    "        if gE == val:\n",
    "            cD_treeg_dic[val] += aE[gE]\n",
    "            \n",
    "for gM in list(aM.keys()):\n",
    "    for val in cD_treeg_dic.keys():\n",
    "        if gM == val:\n",
    "            cD_treeg_dic[val] += aM[gM]\n",
    "\n",
    "for gORF6 in list(aORF6.keys()):\n",
    "    for val in cD_treeg_dic.keys():\n",
    "        if gORF6 == val:\n",
    "            cD_treeg_dic[val] += aORF6[gORF6]\n",
    "            \n",
    "for gORF7a in list(aORF7a.keys()):\n",
    "    for val in cD_treeg_dic.keys():\n",
    "        if gORF7a == val:\n",
    "            cD_treeg_dic[val] += aORF7a[gORF7a]\n",
    "\n",
    "for gORF8 in list(aORF8.keys()):\n",
    "    for val in cD_treeg_dic.keys():\n",
    "        if gORF8 == val:\n",
    "            cD_treeg_dic[val] += aORF8[gORF8]\n",
    "\n",
    "for gN in list(aN.keys()):\n",
    "    for val in cD_treeg_dic.keys():\n",
    "        if gN == val:\n",
    "            cD_treeg_dic[val] += aN[gN]\n",
    "\n",
    "cD_set_treeg = open('SARS_CoV2_rep_mul_align_set.fa', 'w')\n",
    "\n",
    "for g in list(cD_treeg_dic.keys()):\n",
    "    cD_set_treeg.write(\">{}\\n\".format(g))\n",
    "    cD_set_treeg.write(\"{}\\n\".format(cD_treeg_dic[g]))\n",
    "    \n",
    "cD_set_treeg.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1-(10) : Construct phylogenetic tree about nine genes by using mega X\n",
    "# From the phylogenetic tree, the most of viruses in USA is from China\n",
    "# More detailed explanation about infection path from phylogenetic tree is written on my midterm homework word file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2-(1) : Calculate mutation rate\n",
    "# Assume that the reference genome is NC_045512|China|2019-12-00 because all parts of sequence data are same with reference sequence about nine genes\n",
    "\n",
    "rep_align = 'SARS_CoV2_rep_mul_align_set.fa'\n",
    "rep = read_fasta(rep_align)\n",
    "\n",
    "ref_f_coll = rep['NC_045512|China|2019-12-00']\n",
    "\n",
    "count = {name : [0] for name in rep.keys()}\n",
    "N_or_gap_count = {name : [0] for name in rep.keys()}\n",
    "\n",
    "del count['NC_045512|China|2019-12-00']\n",
    "del N_or_gap_count['NC_045512|China|2019-12-00']\n",
    "\n",
    "for name in list(rep.keys()):\n",
    "    if name != 'NC_045512|China|2019-12-00':\n",
    "        for i in range(len(ref_f_coll)):\n",
    "            if rep[name][i] == 'N' or ref_f_coll[i] == 'N' or rep[name][i] == '-' or ref_f_coll[i] == '-':\n",
    "                N_or_gap_count[name][0] += 1\n",
    "            elif rep[name][i] != ref_f_coll[i]:\n",
    "                count[name][0] += 1\n",
    "                \n",
    "# Assume probability of substitution on N(s) base and -(gap) in aligned data is same with probability of substitution on identified bases\n",
    "# Of course, indeed, the frequency of mutation is different along each position and each gene in aligned data\n",
    "# But, despite of this fact, I assume that the probability of mutation is independently with each position\n",
    "# Then, by binomial distribution, the expected number of mutation on N(s) base and -(gap) is L*p\n",
    "# -> L : the total length of N(s) bases and -(gap)s, p : the probability of substitution on N(s) base and -(gap)\n",
    "\n",
    "for k in N_or_gap_count.keys():\n",
    "    sub_len = len(ref_f_coll)-N_or_gap_count[k][0]\n",
    "    prob_m = count[k][0]/sub_len\n",
    "    N_or_gap_count[k][0] *= prob_m\n",
    "\n",
    "Mutation = {name : [] for name in rep.keys()}\n",
    "\n",
    "for key in list(Mutation.keys()):\n",
    "    if key != 'NC_045512|China|2019-12-00':\n",
    "        mutation_count = count[key][0] + round(N_or_gap_count[key][0])\n",
    "        Mutation[key].append(mutation_count)\n",
    "    \n",
    "del Mutation['NC_045512|China|2019-12-00']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22.252475722154607\n"
     ]
    }
   ],
   "source": [
    "# 2-(2) : Calculate mutation rate per year about entire genome\n",
    "\n",
    "date_dic = {name : [] for name in Mutation.keys()}\n",
    "\n",
    "for key in Mutation.keys():\n",
    "    if key != 'MT326173|USA|2020':\n",
    "        date = key.split('|')[2]\n",
    "        days = int(date.split('-')[2])  \n",
    "        if '2019-12' in date:        \n",
    "            date_dic[key].append(days)\n",
    "        elif '2020-01' in date:\n",
    "            days_01 = days + 31\n",
    "            date_dic[key].append(days_01)\n",
    "        elif '2020-02' in date:\n",
    "            days_02 = days + 62\n",
    "            date_dic[key].append(days_02)\n",
    "        elif '2020-03' in date:\n",
    "            days_03 = days + 91\n",
    "            date_dic[key].append(days_03)\n",
    "        elif '2020-04' in date:\n",
    "            days_04 = days + 122\n",
    "            date_dic[key].append(days_04)\n",
    "    else:\n",
    "        date_dic[key].append(112)\n",
    "                    \n",
    "# The date of 'MT326173|USA|2020' assume 2020-3-21 from phylogenetic tree(NJ) and pairwise genetic distance in above code\n",
    "\n",
    "Mutation_rate = {name : [] for name in Mutation.keys()}\n",
    "\n",
    "for key in Mutation.keys():\n",
    "    if key != 'MN908947|China|2019-12-00':\n",
    "        m_rate = (Mutation[key][0]/date_dic[key][0])*365 # Mutation rate per year with assuming that 'one year=365 days'\n",
    "        Mutation_rate[key].append(m_rate)\n",
    "\n",
    "del Mutation_rate['MN908947|China|2019-12-00']\n",
    "\n",
    "sum_mutation_rate = 0\n",
    "for name in Mutation_rate.keys():\n",
    "    sum_mutation_rate += Mutation_rate[name][0]\n",
    "    \n",
    "avg_mutation_rate_per_year = sum_mutation_rate/744\n",
    "print(avg_mutation_rate_per_year)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3 : Is there genes have higher mutation rate?\n",
    "# 3-(1) Define the function that calculate mutation rate per year\n",
    "\n",
    "def cal_m_rate(rep_align):\n",
    "    rep = read_fasta(rep_align)\n",
    "\n",
    "    ref_f_coll = rep['NC_045512|China|2019-12-00']\n",
    "\n",
    "    count = {name : [0] for name in rep.keys()}\n",
    "    N_or_gap_count = {name : [0] for name in rep.keys()}\n",
    "    del count['NC_045512|China|2019-12-00']\n",
    "    del N_or_gap_count['NC_045512|China|2019-12-00']\n",
    "\n",
    "    for name in list(rep.keys()):\n",
    "        if name != 'NC_045512|China|2019-12-00':\n",
    "            for i in range(len(ref_f_coll)):\n",
    "                if rep[name][i] == 'N' or ref_f_coll[i] == 'N' or rep[name][i] == '-' or ref_f_coll[i] == '-':\n",
    "                    N_or_gap_count[name][0] += 1\n",
    "                elif rep[name][i] != ref_f_coll[i]:\n",
    "                    count[name][0] += 1\n",
    "\n",
    "    for k in N_or_gap_count.keys():\n",
    "        sub_len = len(ref_f_coll)-N_or_gap_count[k][0]\n",
    "        prob_m = count[k][0]/sub_len\n",
    "        N_or_gap_count[k][0] *= prob_m\n",
    "\n",
    "    Mutation = {name : [] for name in rep.keys()}\n",
    "\n",
    "    for key in list(Mutation.keys()):\n",
    "        if key != 'NC_045512|China|2019-12-00':\n",
    "            mutation_count = count[key][0] + round(N_or_gap_count[key][0])\n",
    "            Mutation[key].append(mutation_count)\n",
    "    \n",
    "    del Mutation['NC_045512|China|2019-12-00']\n",
    "    \n",
    "    date_dic = {name : [] for name in Mutation.keys()}\n",
    "\n",
    "    for key in Mutation.keys():\n",
    "        if key != 'MT326173|USA|2020':\n",
    "            date = key.split('|')[2]\n",
    "            days = int(date.split('-')[2])  \n",
    "            if '2019-12' in date:        \n",
    "                date_dic[key].append(days)\n",
    "            elif '2020-01' in date:\n",
    "                days_01 = days + 31\n",
    "                date_dic[key].append(days_01)\n",
    "            elif '2020-02' in date:\n",
    "                days_02 = days + 62\n",
    "                date_dic[key].append(days_02)\n",
    "            elif '2020-03' in date:\n",
    "                days_03 = days + 91\n",
    "                date_dic[key].append(days_03)\n",
    "            elif '2020-04' in date:\n",
    "                days_04 = days + 122\n",
    "                date_dic[key].append(days_04)\n",
    "        else:\n",
    "            date_dic[key].append(112)\n",
    "                    \n",
    "\n",
    "    Mutation_rate = {name : [] for name in Mutation.keys()}\n",
    "\n",
    "    for key in Mutation.keys():\n",
    "        if key != 'MN908947|China|2019-12-00':\n",
    "            m_rate = (Mutation[key][0]/date_dic[key][0])*365 \n",
    "            Mutation_rate[key].append(m_rate)\n",
    "\n",
    "    del Mutation_rate['MN908947|China|2019-12-00']\n",
    "\n",
    "    sum_mutation_rate = 0\n",
    "    for name in Mutation_rate.keys():\n",
    "        sum_mutation_rate += Mutation_rate[name][0]\n",
    "    \n",
    "    avg_mutation_rate_per_year = sum_mutation_rate/744\n",
    "    return avg_mutation_rate_per_year"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14.908217558855267\n",
      "2.223296638044254\n",
      "1.6579057890185043\n",
      "0.04539623541727601\n",
      "0.2442921377530046\n",
      "0.15668024938156033\n",
      "0.12571141916870077\n",
      "1.940535160528402\n",
      "0.9968136308366995\n"
     ]
    }
   ],
   "source": [
    "# 3-(2) Mutation rate per year about each gene\n",
    "# > ORF1ab\n",
    "\n",
    "r_ORF1ab = cal_m_rate('SARS_CoV2_ORF1ab_mul_al.fa')\n",
    "print(r_ORF1ab)\n",
    "\n",
    "# > S gene\n",
    "\n",
    "r_S = cal_m_rate('SARS_CoV2_S_mul_al.fa')\n",
    "print(r_S)\n",
    "\n",
    "# > ORF3a\n",
    "\n",
    "r_ORF3a = cal_m_rate('SARS_CoV2_ORF3a_mul_al.fa')\n",
    "print(r_ORF3a)\n",
    "\n",
    "# > E gene\n",
    "\n",
    "r_E = cal_m_rate('SARS_CoV2_E_mul_al.fa')\n",
    "print(r_E)\n",
    "\n",
    "# > M gene\n",
    "\n",
    "r_M = cal_m_rate('SARS_CoV2_M_mul_al.fa')\n",
    "print(r_M)\n",
    "\n",
    "# > ORF6\n",
    "\n",
    "r_ORF6 = cal_m_rate('SARS_CoV2_ORF6_mul_al.fa')\n",
    "print(r_ORF6)\n",
    "\n",
    "# > ORF7a\n",
    "\n",
    "r_ORF7a = cal_m_rate('SARS_CoV2_ORF7a_mul_al.fa')\n",
    "print(r_ORF7a)\n",
    "\n",
    "# > ORF8\n",
    "\n",
    "r_ORF8 = cal_m_rate('SARS_CoV2_ORF8_mul_al.fa')\n",
    "print(r_ORF8)\n",
    "\n",
    "# > N gene\n",
    "\n",
    "r_N = cal_m_rate('SARS_CoV2_N_mul_al.fa')\n",
    "print(r_N)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21349\n",
      "3822\n",
      "828\n",
      "228\n",
      "669\n",
      "186\n",
      "366\n",
      "366\n",
      "1260\n"
     ]
    }
   ],
   "source": [
    "# 3-(3) Can you explain this situation?\n",
    "# Detailed explanation is written on my midterm homework word file\n",
    "\n",
    "# Define the function calculating the length of gene\n",
    "def cal_len(rep_align):\n",
    "    rep = read_fasta(rep_align)\n",
    "\n",
    "    ref_f_coll = rep['NC_045512|China|2019-12-00']\n",
    "    \n",
    "    L_ref = len(ref_f_coll)\n",
    "    \n",
    "    return L_ref\n",
    "    \n",
    "# The result of each gene length\n",
    "L_ORF1ab = cal_len('SARS_CoV2_ORF1ab_mul_al.fa')\n",
    "print(L_ORF1ab)\n",
    "\n",
    "L_S = cal_len('SARS_CoV2_S_mul_al.fa')\n",
    "print(L_S)\n",
    "\n",
    "L_ORF3a = cal_len('SARS_CoV2_ORF3a_mul_al.fa')\n",
    "print(L_ORF3a)\n",
    "\n",
    "L_E = cal_len('SARS_CoV2_E_mul_al.fa')\n",
    "print(L_E)\n",
    "\n",
    "L_M = cal_len('SARS_CoV2_M_mul_al.fa')\n",
    "print(L_M)\n",
    "\n",
    "L_ORF6 = cal_len('SARS_CoV2_ORF6_mul_al.fa')\n",
    "print(L_ORF6)\n",
    "\n",
    "L_ORF7a = cal_len('SARS_CoV2_ORF7a_mul_al.fa')\n",
    "print(L_ORF7a)\n",
    "\n",
    "L_ORF8 = cal_len('SARS_CoV2_ORF8_mul_al.fa')\n",
    "print(L_ORF8)\n",
    "\n",
    "L_N = cal_len('SARS_CoV2_N_mul_al.fa')\n",
    "print(L_N)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
